<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Creating Spatial Annotations â€¢ SPATA2</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Creating Spatial Annotations">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SPATA2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">3.1.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li>
  <a href="../articles/spata-data.html">SPATAData</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Initiating a SPATA2 object</li>
    <li>
      <a href="../articles/initiation-and-preprocessing-customized.html">Initiation &amp; Preprocessing with individual data</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-merfish.html">Initiation &amp; Preprocessing with MERFISH data</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-visium.html">Initiation &amp; Preprocessing with Visium data</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-visium-hd.html">Initiation &amp; Preprocessing with VisiumHD data</a>
    </li>
    <li class="dropdown-header">Basic Manipulation of a SPATA2 object</li>
    <li>
      <a href="../articles/adding-data.html">Adding Data to SPATA2 objects</a>
    </li>
    <li>
      <a href="../articles/package-compatibility.html">Platform Compatibility</a>
    </li>
    <li>
      <a href="../articles/molecular-variables.html">Molecular Variables</a>
    </li>
    <li>
      <a href="../articles/image-handling.html">Image Handling</a>
    </li>
    <li>
      <a href="../articles/spatial-segmentation.html">Spatial Segmentation</a>
    </li>
    <li class="dropdown-header">Implemented Algorithms &amp; Concepts</li>
    <li>
      <a href="../articles/cnv.html">Copy Number Variations</a>
    </li>
    <li>
      <a href="../articles/clustering.html">Clustering</a>
    </li>
    <li>
      <a href="../articles/dimensional-reduction.html">Dimensional Reduction</a>
    </li>
    <li>
      <a href="../articles/dea.html">Differential Expression</a>
    </li>
    <li>
      <a href="../articles/gsea.html">Gene Set Enrichment</a>
    </li>
    <li>
      <a href="../articles/spatial-measures.html">Spatial Measures</a>
    </li>
    <li class="dropdown-header">Spatial Gradient Screening</li>
    <li>
      <a href="../articles/creating-spatial-annotations.html">Creating Spatial Annotations</a>
    </li>
    <li>
      <a href="../articles/using-spatial-annotations.html">Using Spatial Annotations</a>
    </li>
    <li>
      <a href="../articles/spatial-annotation-screening.html">Spatial Annotation Screening</a>
    </li>
    <li>
      <a href="../articles/creating-spatial-trajectories.html">Creating Spatial Trajectories</a>
    </li>
    <li>
      <a href="../articles/spatial-trajectory-screening.html">Spatial Trajectory Screening</a>
    </li>
    <li class="dropdown-header">Visualization</li>
    <li>
      <a href="../articles/plotting-surface.html">Surface Plotting</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/newslog.html">Newslog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Creating Spatial Annotations</h1>
            
      

      <div class="hidden name"><code>creating-spatial-annotations.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load required packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://themilolab.com/" class="external-link">SPATA2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># load SPATA2 inbuilt example data</span></span>
<span><span class="va">object_t313</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadExampleObject.html">loadExampleObject</a></span><span class="op">(</span><span class="st">"UKF313T"</span>, process <span class="op">=</span> <span class="cn">TRUE</span>, meta <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">object_t275</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadExampleObject.html">loadExampleObject</a></span><span class="op">(</span><span class="st">"UKF275T"</span>, process <span class="op">=</span> <span class="cn">TRUE</span>, meta <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="spatial-annotations-in-general">1. Spatial Annotations in general<a class="anchor" aria-label="anchor" href="#spatial-annotations-in-general"></a>
</h2>
<p>Spatial Annotations represent annotations for spatial data. The
concept allows users to define and store polygons that outline areas of
interest within images or datasets. The outlines, in turn, can be used
as spatial references to visualize feature expression as a function of
distance to areas that are potential biological forces. Generally
speaking, a spatial annotation simplay require a polygon. SPATA2
implements three concepts of Spatial Annotations: (Spatial) Image
Annotations, (Spatial) Group Annotations and (Spatial) Numeric
Annotations.</p>
</div>
<div class="section level2">
<h2 id="image-annotations">2. Image Annotations<a class="anchor" aria-label="anchor" href="#image-annotations"></a>
</h2>
<p>This section uses the sample <em>UKF313T</em> as an example.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot image on the left</span></span>
<span><span class="fu"><a href="../reference/plotImage.html">plotImage</a></span><span class="op">(</span><span class="va">object_t313</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># uses the results of computeMetaFeatures(), count distribution is heavily right skewed -&gt; transform</span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object_t313</span>, color_by <span class="op">=</span> <span class="st">"n_counts_gene"</span>, transform_with <span class="op">=</span> <span class="va">log10</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"counts\n(log10)"</span><span class="op">)</span></span></code></pre></div>
<p><img src="creating-spatial-annotations_files/figure-html/unnamed-chunk-3-1.png" width="50%"><img src="creating-spatial-annotations_files/figure-html/unnamed-chunk-3-2.png" width="50%">
Image annotations are represented by the S4 class
<code>ImageAnnotation</code>, which is designed to capture spatial
annotations by outlining areas of interest on images. It provides a
flexible framework for creating annotations that visually highlight
specific regions within images, such as histological structures,
cellular patterns, or other histo-morphological features in images. It
can be interactively created with the
<code><a href="../reference/createImageAnnotations.html">createImageAnnotations()</a></code> function. This function lets you
access an interactive application in which you can encircle the
structure or area you want to annotate.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># access an interface to interact with the image</span></span>
<span><span class="va">object_t313</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createImageAnnotations.html">createImageAnnotations</a></span><span class="op">(</span><span class="va">object_t313</span><span class="op">)</span></span></code></pre></div>
<video width="640" height="480" controls><source src="../reference/figures/create_image_annotation_t313.mp4" type="video/mp4" srcset="NA"></source></video><p>
</p>
<p>The left plot <em>Interaction</em> is where the magic happens. The
right plot is used for orientation if you want to zoom in and out.
Double click on the left image to start the drawing process. Double
click again to stop drawing. If you are using drawing mode
<em>Single</em> click on â€˜Highlightâ€™ to highlight the encircled area,
enter the tags you want to tag the annotation with, enter the ID with
which you want to name the annotation and click on â€˜Add Image
Annotationâ€™. If you are in drawing mode <em>Multiple</em> stopping the
drawing immediately highlights the encircled area. This allows to
quickly encircle multiple structures of the same kind that are tagged
with the same tags (e.g.Â multiple small vessel). The tab on the right
called â€˜Added Image Annotationsâ€™ allows to visualize all image
annotations saved so far. Make sure to click on â€˜Close applicationâ€™ to
return the <code>SPATA2</code> object containing the results.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># results are stored in the processed object that can be downloaded </span></span>
<span><span class="fu"><a href="../reference/plotSpatialAnnotations.html">plotSpatialAnnotations</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">object_t313</span>,</span>
<span>  ids <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"necrotic_area"</span>, <span class="st">"necrotic_edge"</span>, <span class="st">"necrotic_edge2"</span><span class="op">)</span>, </span>
<span>  nrow <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  unit <span class="op">=</span> <span class="st">"px"</span>, </span>
<span>  fill <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/alpha.html" class="external-link">alpha</a></span><span class="op">(</span><span class="st">"orange"</span>, <span class="fl">0.25</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="creating-spatial-annotations_files/figure-html/unnamed-chunk-6-1.png" width="100%"></p>
<p>Note that the image annotations <em>necrotic_edge</em>
<em>necrotic_edge2</em> have been postprocessed with
<code><a href="../reference/mergeWithTissueOutline.html">mergeWithTissueOutline()</a></code> to prevent any transgression of
the tissue edge.</p>
</div>
<div class="section level2">
<h2 id="numeric-annotations">3. Numeric Annotations<a class="anchor" aria-label="anchor" href="#numeric-annotations"></a>
</h2>
<p>Numeric annotations are designed to represent the spatial extent of
data points, such as cells or barcoded spots, by filtering and outlining
them according to their values for a specific numeric variable. This is
particularly suitable for creating annotations that highlight areas of
interest based on continuous characteristics like gene expression or
other numeric attributes derived from spatial multi-omic datasets.
SPATA2 allows to annotate space based on numeric variables automatically
using <code><a href="../reference/createNumericAnnotations.html">createNumericAnnotations()</a></code>. For more details on how
to manipulate the way the areas are annotated, please refer to the
documentation via <code>createNumericAnnotations</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># left plot</span></span>
<span><span class="fu"><a href="../reference/plotImage.html">plotImage</a></span><span class="op">(</span><span class="va">object_t275</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># right plot</span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object_t275</span>, color_by <span class="op">=</span> <span class="st">"HM_HYPOXIA"</span>, pt_clrsp <span class="op">=</span> <span class="st">"plasma"</span><span class="op">)</span> </span></code></pre></div>
<p><img src="creating-spatial-annotations_files/figure-html/unnamed-chunk-7-1.png" width="50%"><img src="creating-spatial-annotations_files/figure-html/unnamed-chunk-7-2.png" width="50%"></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">object_t275</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/createNumericAnnotations.html">createNumericAnnotations</a></span><span class="op">(</span></span>
<span>    object <span class="op">=</span> <span class="va">object_t275</span>, </span>
<span>    variable <span class="op">=</span> <span class="st">"HM_HYPOXIA"</span>, </span>
<span>    threshold <span class="op">=</span> <span class="st">"kmeans_high"</span>,</span>
<span>    inner_borders <span class="op">=</span> <span class="cn">TRUE</span>, <span class="co"># default, allows "holes" inside the annotation</span></span>
<span>    id <span class="op">=</span> <span class="st">"hypoxic_ann"</span>, </span>
<span>    force1 <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># all spatial annotation IDs</span></span>
<span><span class="fu"><a href="../reference/getSpatAnnIds.html">getSpatAnnIds</a></span><span class="op">(</span><span class="va">object_t275</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "vessel1"     "vessel2"     "vessel3"     "img_ann_1"   "hypoxic_ann"</span></span></code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># subset with their class</span></span>
<span><span class="fu"><a href="../reference/getSpatAnnIds.html">getSpatAnnIds</a></span><span class="op">(</span><span class="va">object_t275</span>, class <span class="op">=</span> <span class="st">"Numeric"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "hypoxic_ann"</span></span></code></pre>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot the outline</span></span>
<span><span class="va">hypoxic_ann_outline</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/ggpLayerSpatAnnOutline.html">ggpLayerSpatAnnOutline</a></span><span class="op">(</span><span class="va">object_t275</span>, ids <span class="op">=</span> <span class="st">"hypoxic_ann"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># left plot</span></span>
<span><span class="fu"><a href="../reference/plotImage.html">plotImage</a></span><span class="op">(</span><span class="va">object_t275</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="va">hypoxic_ann_outline</span></span>
<span></span>
<span><span class="co"># right plot</span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object_t275</span>, color_by <span class="op">=</span> <span class="st">"HM_HYPOXIA"</span>, pt_clrsp <span class="op">=</span> <span class="st">"plasma"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="va">hypoxic_ann_outline</span></span></code></pre></div>
<p><img src="creating-spatial-annotations_files/figure-html/unnamed-chunk-8-1.png" width="50%"><img src="creating-spatial-annotations_files/figure-html/unnamed-chunk-8-2.png" width="50%"></p>
<p>Note the difference between <code>inner_borders = TRUE</code> and
<code>inner_borders = FALSE</code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">object_t275</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/createNumericAnnotations.html">createNumericAnnotations</a></span><span class="op">(</span></span>
<span>    object <span class="op">=</span> <span class="va">object_t275</span>, </span>
<span>    variable <span class="op">=</span> <span class="st">"HM_HYPOXIA"</span>, </span>
<span>    threshold <span class="op">=</span> <span class="st">"kmeans_high"</span>,</span>
<span>    inner_borders <span class="op">=</span> <span class="cn">FALSE</span>, <span class="co"># no holes, only one outer outline</span></span>
<span>    id <span class="op">=</span> <span class="st">"hypoxic_ann_nh"</span>, </span>
<span>    force1 <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># left plot</span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object_t275</span>, color_by <span class="op">=</span> <span class="st">"HM_HYPOXIA"</span>, pt_clrsp <span class="op">=</span> <span class="st">"plasma"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="va">hypoxic_ann_outline</span></span>
<span></span>
<span><span class="co"># right plot</span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object_t275</span>, color_by <span class="op">=</span> <span class="st">"HM_HYPOXIA"</span>, pt_clrsp <span class="op">=</span> <span class="st">"plasma"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="../reference/ggpLayerSpatAnnOutline.html">ggpLayerSpatAnnOutline</a></span><span class="op">(</span><span class="va">object_t275</span>, ids <span class="op">=</span> <span class="st">"hypoxic_ann_nh"</span><span class="op">)</span></span></code></pre></div>
<p><img src="creating-spatial-annotations_files/figure-html/unnamed-chunk-9-1.png" width="50%"><img src="creating-spatial-annotations_files/figure-html/unnamed-chunk-9-2.png" width="50%"></p>
</div>
<div class="section level2">
<h2 id="group-annotations">4. Group Annotations<a class="anchor" aria-label="anchor" href="#group-annotations"></a>
</h2>
<p>Group Annotations are designed to represent the spatial extent of
data points, such as cells or barcoded spots, by filtering and outlining
them based on predefined groups. This allows for the creation of
annotations that highlight specific spatial clusters, areas, or patterns
identified through grouping techniques. It provides a means to focus on
regions of interest within spatial multi-omic datasets using predefined
categorizations. To create such spatial annotations use the function
<code><a href="../reference/createGroupAnnotations.html">createGroupAnnotations()</a></code>. For more details on how to
manipulate the way the areas are annotated, please refer to the
documentation via <code>createGroupAnnotations</code>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot sample </span></span>
<span><span class="fu"><a href="../reference/plotImage.html">plotImage</a></span><span class="op">(</span><span class="va">object_t275</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object_t275</span>, color_by <span class="op">=</span> <span class="st">"bayes_space"</span>, pt_clrp <span class="op">=</span> <span class="st">"jco"</span><span class="op">)</span></span></code></pre></div>
<p><img src="creating-spatial-annotations_files/figure-html/unnamed-chunk-10-1.png" width="50%"><img src="creating-spatial-annotations_files/figure-html/unnamed-chunk-10-2.png" width="50%"></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create the annotation</span></span>
<span><span class="va">object_t275</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/createGroupAnnotations.html">createGroupAnnotations</a></span><span class="op">(</span></span>
<span>    object <span class="op">=</span> <span class="va">object_t275</span>, </span>
<span>    grouping <span class="op">=</span> <span class="st">"bayes_space"</span>, </span>
<span>    group <span class="op">=</span> <span class="st">"B4"</span>,</span>
<span>    id <span class="op">=</span> <span class="st">"bspace4"</span>, </span>
<span>    force1 <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># all spatial annotation IDs</span></span>
<span><span class="fu"><a href="../reference/getSpatAnnIds.html">getSpatAnnIds</a></span><span class="op">(</span><span class="va">object_t275</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "vessel1"        "vessel2"        "vessel3"        "img_ann_1"     </span></span>
<span><span class="co">## [5] "hypoxic_ann"    "hypoxic_ann_nh" "bspace4"</span></span></code></pre>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># subset with their class</span></span>
<span><span class="fu"><a href="../reference/getSpatAnnIds.html">getSpatAnnIds</a></span><span class="op">(</span><span class="va">object_t275</span>, class <span class="op">=</span> <span class="st">"Numeric"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "hypoxic_ann"    "hypoxic_ann_nh"</span></span></code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot the outline</span></span>
<span><span class="va">cluster_ann_outline</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ggpLayerSpatAnnOutline.html">ggpLayerSpatAnnOutline</a></span><span class="op">(</span><span class="va">object_t275</span>, ids <span class="op">=</span> <span class="st">"bspace4"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotImage.html">plotImage</a></span><span class="op">(</span><span class="va">object_t275</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="va">cluster_ann_outline</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object_t275</span>, color_by <span class="op">=</span> <span class="st">"bayes_space"</span>, pt_clrp <span class="op">=</span> <span class="st">"jco"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="va">cluster_ann_outline</span></span></code></pre></div>
<p><img src="creating-spatial-annotations_files/figure-html/unnamed-chunk-11-1.png" width="50%"><img src="creating-spatial-annotations_files/figure-html/unnamed-chunk-11-2.png" width="50%"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Jan Kueckelhaus, Dieter-Henrik Heiland.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
