<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Differential Expression Analysis (DEA) â€¢ SPATA2</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Differential Expression Analysis (DEA)">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SPATA2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">3.1.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li>
  <a href="../articles/spata-data.html">SPATAData</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Initiating a SPATA2 object</li>
    <li>
      <a href="../articles/initiation-and-preprocessing-customized.html">Initiation &amp; Preprocessing with individual data</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-merfish.html">Initiation &amp; Preprocessing with MERFISH data</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-visium.html">Initiation &amp; Preprocessing with Visium data</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-visium-hd.html">Initiation &amp; Preprocessing with VisiumHD data</a>
    </li>
    <li class="dropdown-header">Basic Manipulation of a SPATA2 object</li>
    <li>
      <a href="../articles/adding-data.html">Adding Data to SPATA2 objects</a>
    </li>
    <li>
      <a href="../articles/package-compatibility.html">Platform Compatibility</a>
    </li>
    <li>
      <a href="../articles/molecular-variables.html">Molecular Variables</a>
    </li>
    <li>
      <a href="../articles/image-handling.html">Image Handling</a>
    </li>
    <li>
      <a href="../articles/spatial-segmentation.html">Spatial Segmentation</a>
    </li>
    <li class="dropdown-header">Implemented Algorithms &amp; Concepts</li>
    <li>
      <a href="../articles/cnv.html">Copy Number Variations</a>
    </li>
    <li>
      <a href="../articles/clustering.html">Clustering</a>
    </li>
    <li>
      <a href="../articles/dimensional-reduction.html">Dimensional Reduction</a>
    </li>
    <li>
      <a href="../articles/dea.html">Differential Expression</a>
    </li>
    <li>
      <a href="../articles/gsea.html">Gene Set Enrichment</a>
    </li>
    <li>
      <a href="../articles/spatial-measures.html">Spatial Measures</a>
    </li>
    <li class="dropdown-header">Spatial Gradient Screening</li>
    <li>
      <a href="../articles/creating-spatial-annotations.html">Creating Spatial Annotations</a>
    </li>
    <li>
      <a href="../articles/using-spatial-annotations.html">Using Spatial Annotations</a>
    </li>
    <li>
      <a href="../articles/spatial-annotation-screening.html">Spatial Annotation Screening</a>
    </li>
    <li>
      <a href="../articles/creating-spatial-trajectories.html">Creating Spatial Trajectories</a>
    </li>
    <li>
      <a href="../articles/spatial-trajectory-screening.html">Spatial Trajectory Screening</a>
    </li>
    <li class="dropdown-header">Visualization</li>
    <li>
      <a href="../articles/plotting-surface.html">Surface Plotting</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/newslog.html">Newslog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Differential Expression Analysis (DEA)</h1>
            
      

      <div class="hidden name"><code>dea.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">1. Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Differential expression analysis (DEA) aims to discover quantitative
changes in gene expression levels between defined experimental groups.
Grouping information is stored in form of grouping variables in the meta
data of <code>SPATA2</code> object. This includes all SPATA2 intern
generated grouping options such as <a href="spatial-segmentation.html">spatial segmentation</a> and <a href="clustering.html">clustering</a> as well as any other grouping
variable that has been added via <code><a href="../reference/addFeatures.html">addFeatures()</a></code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load required packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://themilolab.com/" class="external-link">SPATA2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># load SPATA2 inbuilt example data</span></span>
<span><span class="va">object_t269</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadExampleObject.html">loadExampleObject</a></span><span class="op">(</span><span class="st">"UKF269T"</span>, process <span class="op">=</span> <span class="cn">TRUE</span>, meta <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot histology </span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object_t269</span>, color_by <span class="op">=</span> <span class="st">"histology"</span>, pt_clrp <span class="op">=</span> <span class="st">"npg"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot bayes space cluster</span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object_t269</span>, color_by <span class="op">=</span> <span class="st">"bayes_space"</span>, pt_clrp <span class="op">=</span> <span class="st">"uc"</span><span class="op">)</span></span></code></pre></div>
<p><img src="dea_files/figure-html/unnamed-chunk-2-1.png" width="50%"><img src="dea_files/figure-html/unnamed-chunk-2-2.png" width="50%"></p>
<p>The function <code><a href="../reference/getGroupingOptions.html">getGroupingOptions()</a></code> returns the variable
names based on which DEA can be conducted. This vignette conducts
differential expression analysis based on the histological grouping. You
can exchange <em>â€˜histologyâ€™</em> with <em>â€˜bayes_spaceâ€™</em> whenever
the grouping is specified to use the clustering results.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/getGroupingOptions.html">getGroupingOptions</a></span><span class="op">(</span><span class="va">object_t269</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##            factor            factor            factor            factor </span></span>
<span><span class="co">##  "tissue_section" "seurat_clusters"       "histology"     "bayes_space"</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="running-the-analysis">2. Running the analysis<a class="anchor" aria-label="anchor" href="#running-the-analysis"></a>
</h2>
<p>SPATA2 uses the function <code><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html" class="external-link">Seurat::FindAllMarkers()</a></code> for
differential expression analysis. Itâ€™s output is a data.frame in which
each row corresponds to a gene that turned out to be a marker gene for
one of the identity groups. Additional variables provide information
about itâ€™s p-value, adjusted p-value, logFold change etc.
<code><a href="../reference/runDEA.html">runDEA()</a></code> does not alter the output but stores it in the
<code>SPATA2</code> object.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">object_t269</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runDEA.html">runDEA</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">object_t269</span>, across <span class="op">=</span> <span class="st">"histology"</span>, method <span class="op">=</span> <span class="st">"wilcox"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="extracting-results">3. Extracting results<a class="anchor" aria-label="anchor" href="#extracting-results"></a>
</h2>
<p>There are two main functions with which you can manually extract the
DEA results desired. First, <code><a href="../reference/getDeaResultsDf.html">getDeaResultsDf()</a></code> returns the
original resulting data.frame of <code><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html" class="external-link">Seurat::FindAllMarkers()</a></code>.
<code><a href="../reference/getDeaResultsDf.html">getDeaGenes()</a></code> returns a vector of gene names.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># extract the complete data.frame</span></span>
<span><span class="va">dea_df</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/getDeaResultsDf.html">getDeaResultsDf</a></span><span class="op">(</span></span>
<span>    object <span class="op">=</span> <span class="va">object_t269</span>, </span>
<span>    across <span class="op">=</span> <span class="st">"histology"</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">dea_df</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 22298</span></span></code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">dea_df</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 6 Ã— 7</span></span></span>
<span><span class="co">##      p_val avg_log2FC pct.1 pct.2 p_val_adj histology gene     </span></span>
<span><span class="co">##      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> 3.26<span style="color: #949494;">e</span><span style="color: #BB0000;">-17</span>       7.54 0.037     0  4.89<span style="color: #949494;">e</span><span style="color: #BB0000;">-13</span> tumor     SAA2     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> 7.28<span style="color: #949494;">e</span><span style="color: #BB0000;">-23</span>       7.07 0.05      0  1.09<span style="color: #949494;">e</span><span style="color: #BB0000;">-18</span> tumor     NODAL    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> 1.55<span style="color: #949494;">e</span><span style="color: #BB0000;">-19</span>       6.85 0.042     0  2.33<span style="color: #949494;">e</span><span style="color: #BB0000;">-15</span> tumor     MPZL2    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> 7.17<span style="color: #949494;">e</span><span style="color: #BB0000;">-19</span>       6.69 0.041     0  1.08<span style="color: #949494;">e</span><span style="color: #BB0000;">-14</span> tumor     LINC02308</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span> 7.09<span style="color: #949494;">e</span><span style="color: #BB0000;">-18</span>       6.61 0.038     0  1.06<span style="color: #949494;">e</span><span style="color: #BB0000;">-13</span> tumor     CD48     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span> 3.26<span style="color: #949494;">e</span><span style="color: #BB0000;">-17</span>       6.56 0.037     0  4.89<span style="color: #949494;">e</span><span style="color: #BB0000;">-13</span> tumor     HOXA-AS2</span></span></code></pre>
<p>Using the arguments <code>across_subset</code>, <code>min_lfc</code>,
<code>n_highest_lfc</code>, <code>max_adj_pval</code>,
<code>n_lowest_pval</code> the output of the function can be adjusted to
specific questions.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># e.g. top 10 genes for histology area 'tumor' </span></span>
<span><span class="fu"><a href="../reference/getDeaResultsDf.html">getDeaResultsDf</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">object_t269</span>, </span>
<span>  across <span class="op">=</span> <span class="st">"histology"</span>,</span>
<span>  across_subset <span class="op">=</span> <span class="st">"transition"</span>, <span class="co"># the group name(s) of interest,</span></span>
<span>  n_highest_lfc <span class="op">=</span> <span class="fl">10</span>, <span class="co"># top ten genes</span></span>
<span>  max_adj_pval <span class="op">=</span> <span class="fl">0.01</span> <span class="co"># pval must be lower or equal than 0.01</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 10 Ã— 7</span></span></span>
<span><span class="co">##        p_val avg_log2FC pct.1 pct.2 p_val_adj histology  gene     </span></span>
<span><span class="co">##        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> 8.53<span style="color: #949494;">e</span><span style="color: #BB0000;">-  8</span>       7.98 0.01  0     1.28<span style="color: #949494;">e</span><span style="color: #BB0000;">-  3</span> transition CPEB1-AS1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> 1.21<span style="color: #949494;">e</span><span style="color: #BB0000;">- 11</span>       4.59 0.042 0.006 1.82<span style="color: #949494;">e</span><span style="color: #BB0000;">-  7</span> transition SGK2     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> 8.16<span style="color: #949494;">e</span><span style="color: #BB0000;">- 30</span>       4.40 0.111 0.016 1.22<span style="color: #949494;">e</span><span style="color: #BB0000;">- 25</span> transition GPIHBP1  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> 2.18<span style="color: #949494;">e</span><span style="color: #BB0000;">- 20</span>       4.34 0.075 0.011 3.28<span style="color: #949494;">e</span><span style="color: #BB0000;">- 16</span> transition FAM95C   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> 1.00<span style="color: #949494;">e</span><span style="color: #BB0000;">- 10</span>       4.07 0.04  0.006 1.50<span style="color: #949494;">e</span><span style="color: #BB0000;">-  6</span> transition HOXD1    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> 6.57<span style="color: #949494;">e</span><span style="color: #BB0000;">- 72</span>       4.03 0.303 0.057 9.86<span style="color: #949494;">e</span><span style="color: #BB0000;">- 68</span> transition GJB1     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> 6.05<span style="color: #949494;">e</span><span style="color: #BB0000;">- 12</span>       3.90 0.054 0.011 9.08<span style="color: #949494;">e</span><span style="color: #BB0000;">-  8</span> transition SMIM6    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> 1.65<span style="color: #949494;">e</span><span style="color: #BB0000;">-  8</span>       3.70 0.061 0.018 2.48<span style="color: #949494;">e</span><span style="color: #BB0000;">-  4</span> transition PIEZO2   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> 4.50<span style="color: #949494;">e</span><span style="color: #BB0000;">- 15</span>       3.69 0.088 0.022 6.75<span style="color: #949494;">e</span><span style="color: #BB0000;">- 11</span> transition LRP2     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> 2.39<span style="color: #949494;">e</span><span style="color: #BB0000;">-168</span>       3.68 0.72  0.2   3.58<span style="color: #949494;">e</span><span style="color: #BB0000;">-164</span> transition MAG</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="visualize-results">4. Visualize results<a class="anchor" aria-label="anchor" href="#visualize-results"></a>
</h2>
<div class="section level3">
<h3 id="heatmaps">4.1 Heatmaps<a class="anchor" aria-label="anchor" href="#heatmaps"></a>
</h3>
<p><code><a href="../reference/plotDeaHeatmap.html">plotDeaHeatmap()</a></code> visualizes the results of DEA by using
to the results you would extract with
<code><a href="../reference/getDeaResultsDf.html">getDeaResultsDf()</a></code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hm</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/plotDeaHeatmap.html">plotDeaHeatmap</a></span><span class="op">(</span></span>
<span>    object <span class="op">=</span> <span class="va">object_t269</span>, </span>
<span>    across <span class="op">=</span> <span class="st">"histology"</span>,</span>
<span>    clrp <span class="op">=</span> <span class="st">"npg"</span>,</span>
<span>    n_highest_lfc <span class="op">=</span> <span class="fl">10</span>, <span class="co"># subset genes</span></span>
<span>    n_bcs <span class="op">=</span> <span class="fl">100</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="va">hm</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="dotplots">4.2 Dotplots<a class="anchor" aria-label="anchor" href="#dotplots"></a>
</h3>
<p><code><a href="../reference/plotDeaDotPlot.html">plotDeaDotPlot()</a></code> visualizes the results of DEA either by
groupâ€¦</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotDeaDotPlot.html">plotDeaDotPlot</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">object_t269</span>, </span>
<span>  across <span class="op">=</span> <span class="st">"histology"</span>,</span>
<span>  across_subset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"tumor"</span>, <span class="st">"infiltrated"</span><span class="op">)</span>, <span class="co"># specify single groups if desired</span></span>
<span>  n_highest_lfc <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  by_group <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>  scales <span class="op">=</span> <span class="st">"free_y"</span>, </span>
<span>  nrow <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="dea_files/figure-html/unnamed-chunk-8-1.png" width="100%"></p>
<p>â€¦ or with all groups together.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotDeaDotPlot.html">plotDeaDotPlot</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">object_t269</span>, </span>
<span>  across <span class="op">=</span> <span class="st">"histology"</span>,</span>
<span>  color_by <span class="op">=</span> <span class="st">"histology"</span>,</span>
<span>  pt_clrp <span class="op">=</span> <span class="st">"npg"</span>,</span>
<span>  size_by <span class="op">=</span> <span class="st">"avg_log2FC"</span>,</span>
<span>  n_highest_lfc <span class="op">=</span> <span class="fl">5</span>, </span>
<span>  by_group <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="dea_files/figure-html/unnamed-chunk-9-1.png" width="100%"></p>
</div>
<div class="section level3">
<h3 id="box--and-violinplots">4.3 Box- and Violinplots<a class="anchor" aria-label="anchor" href="#box--and-violinplots"></a>
</h3>
<p>There are additional ways to visualize the results of your DEA. As
with almost all plotting functions in <strong>SPATA2</strong> a vector
of gene names is necessary for the function to know which genes to plot.
<code><a href="../reference/getDeaResultsDf.html">getDeaGenes()</a></code> is the second function to extract DEA results
and a wrapper around <code><a href="../reference/getDeaResultsDf.html">getDeaResultsDf()</a></code> that returns a
vector gene names.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genes_of_interest</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="../reference/getDeaResultsDf.html">getDeaGenes</a></span><span class="op">(</span></span>
<span>    object <span class="op">=</span> <span class="va">object_t269</span>,</span>
<span>    across <span class="op">=</span> <span class="st">"histology"</span>, <span class="co"># the grouping variable</span></span>
<span>    method_de <span class="op">=</span> <span class="st">"wilcox"</span>, <span class="co"># the method with which the results were computed</span></span>
<span>    n_highest_lfc <span class="op">=</span> <span class="fl">10</span>, </span>
<span>    max_adj_pval <span class="op">=</span> <span class="fl">0.001</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">genes_of_interest</span><span class="op">)</span> <span class="co"># first six</span></span></code></pre></div>
<pre><code><span><span class="co">##       tumor       tumor       tumor       tumor       tumor       tumor </span></span>
<span><span class="co">##      "SAA2"     "NODAL"     "MPZL2" "LINC02308"      "CD48"  "HOXA-AS2"</span></span></code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="va">genes_of_interest</span><span class="op">)</span> <span class="co"># last six</span></span></code></pre></div>
<pre><code><span><span class="co">## infiltrated infiltrated infiltrated infiltrated infiltrated infiltrated </span></span>
<span><span class="co">##     "HTR5A"     "CRHR2"    "CACNG2"    "TRIM54"     "ACTC1"    "KRT222"</span></span></code></pre>
<p>A vector of gene names as returned by <code><a href="../reference/getDeaResultsDf.html">getDeaGenes()</a></code> is a
perfectly valid input for other plotting functions.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">top_5_markers_269</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/getDeaResultsDf.html">getDeaGenes</a></span><span class="op">(</span></span>
<span>    object <span class="op">=</span> <span class="va">object_t269</span>,</span>
<span>    across <span class="op">=</span> <span class="st">"histology"</span>,</span>
<span>    n_lowest_pval <span class="op">=</span> <span class="fl">5</span>, </span>
<span>    min_lfc <span class="op">=</span> <span class="fl">0.1</span> <span class="co"># set min_lfc! else downregulated genes are included</span></span>
<span>    <span class="op">)</span> </span>
<span></span>
<span><span class="va">top_5_markers_269</span></span></code></pre></div>
<pre><code><span><span class="co">##       tumor       tumor       tumor       tumor       tumor  transition </span></span>
<span><span class="co">##    "IGFBP2" "LINC01445"     "SOCS2"       "TNC"     "MATN2"       "MAG" </span></span>
<span><span class="co">##  transition  transition  transition  transition infiltrated infiltrated </span></span>
<span><span class="co">##      "MOBP"        "TF"      "PLP1"       "MBP"       "CCK"     "VSNL1" </span></span>
<span><span class="co">## infiltrated infiltrated infiltrated </span></span>
<span><span class="co">##      "NRGN"     "OLFM1"   "SLC17A7"</span></span></code></pre>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot results for t269</span></span>
<span><span class="fu"><a href="../reference/plotBoxplot.html">plotBoxplot</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">object_t269</span>,</span>
<span>  variables <span class="op">=</span> <span class="va">top_5_markers_269</span>,</span>
<span>  across <span class="op">=</span> <span class="st">"histology"</span>, </span>
<span>  clrp <span class="op">=</span> <span class="st">"npg"</span>,</span>
<span>  nrow <span class="op">=</span> <span class="fl">3</span> </span>
<span>  <span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>, axis.ticks.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="../reference/legendBottom.html">legendTop</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="dea_files/figure-html/unnamed-chunk-11-1.png" width="100%"></p>
</div>
<div class="section level3">
<h3 id="surface-plots">4.4 Surface plots<a class="anchor" aria-label="anchor" href="#surface-plots"></a>
</h3>
<p><code><a href="../reference/getDeaResultsDf.html">getDeaGenes()</a></code> complements
<code><a href="../reference/plotSurfaceComparison.html">plotSurfaceComparison()</a></code>.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># top 9 markers for transition area</span></span>
<span><span class="va">transition_markers</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/getDeaResultsDf.html">getDeaGenes</a></span><span class="op">(</span><span class="va">object_t269</span>, across <span class="op">=</span> <span class="st">"histology"</span>, across_subset <span class="op">=</span> <span class="st">"transition"</span>, n_lowest_pval <span class="op">=</span> <span class="fl">9</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotSurfaceComparison.html">plotSurfaceComparison</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">object_t269</span>, </span>
<span>  color_by <span class="op">=</span> <span class="va">transition_markers</span>,</span>
<span>  pt_clrsp <span class="op">=</span> <span class="fu"><a href="../reference/color_vector.html">color_vector</a></span><span class="op">(</span><span class="st">"npg"</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, <span class="co"># plot cluster color against white</span></span>
<span>  outline <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  nrow <span class="op">=</span> <span class="fl">3</span></span>
<span>  <span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="../reference/legendBottom.html">legendNone</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="dea_files/figure-html/unnamed-chunk-12-1.png" width="100%"></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Jan Kueckelhaus, Dieter-Henrik Heiland.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
