<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Initiation &amp; Preprocessing: MERFISH â€¢ SPATA2</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Initiation &amp; Preprocessing: MERFISH">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SPATA2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">3.1.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li>
  <a href="../articles/spata-data.html">SPATAData</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Initiating a SPATA2 object</li>
    <li>
      <a href="../articles/initiation-and-preprocessing-customized.html">Initiation &amp; Preprocessing with individual data</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-merfish.html">Initiation &amp; Preprocessing with MERFISH data</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-visium.html">Initiation &amp; Preprocessing with Visium data</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-visium-hd.html">Initiation &amp; Preprocessing with VisiumHD data</a>
    </li>
    <li class="dropdown-header">Basic Manipulation of a SPATA2 object</li>
    <li>
      <a href="../articles/adding-data.html">Adding Data to SPATA2 objects</a>
    </li>
    <li>
      <a href="../articles/package-compatibility.html">Platform Compatibility</a>
    </li>
    <li>
      <a href="../articles/molecular-variables.html">Molecular Variables</a>
    </li>
    <li>
      <a href="../articles/image-handling.html">Image Handling</a>
    </li>
    <li>
      <a href="../articles/spatial-segmentation.html">Spatial Segmentation</a>
    </li>
    <li class="dropdown-header">Implemented Algorithms &amp; Concepts</li>
    <li>
      <a href="../articles/cnv.html">Copy Number Variations</a>
    </li>
    <li>
      <a href="../articles/clustering.html">Clustering</a>
    </li>
    <li>
      <a href="../articles/dimensional-reduction.html">Dimensional Reduction</a>
    </li>
    <li>
      <a href="../articles/dea.html">Differential Expression</a>
    </li>
    <li>
      <a href="../articles/gsea.html">Gene Set Enrichment</a>
    </li>
    <li>
      <a href="../articles/spatial-measures.html">Spatial Measures</a>
    </li>
    <li class="dropdown-header">Spatial Gradient Screening</li>
    <li>
      <a href="../articles/creating-spatial-annotations.html">Creating Spatial Annotations</a>
    </li>
    <li>
      <a href="../articles/using-spatial-annotations.html">Using Spatial Annotations</a>
    </li>
    <li>
      <a href="../articles/spatial-annotation-screening.html">Spatial Annotation Screening</a>
    </li>
    <li>
      <a href="../articles/creating-spatial-trajectories.html">Creating Spatial Trajectories</a>
    </li>
    <li>
      <a href="../articles/spatial-trajectory-screening.html">Spatial Trajectory Screening</a>
    </li>
    <li class="dropdown-header">Visualization</li>
    <li>
      <a href="../articles/plotting-surface.html">Surface Plotting</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/newslog.html">Newslog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Initiation &amp; Preprocessing: MERFISH</h1>
            
      

      <div class="hidden name"><code>initiation-and-preprocessing-merfish.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">1. Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This tutorial demonstrates how to load and preprocess MERFISH data in
SPATA2 using the <code><a href="../reference/initiateSpataObjectMERFISH.html">initiateSpataObjectMERFISH()</a></code> function. The
required files are a cell-by-gene matrix (<em>cell_by_gene.csv</em>) and
cell metadata (<em>cell_metadata.csv</em>) from the MERFISH output
folder. The function can be adjusted for custom files, see
<code>?initiateSpataObjectMERFISH()</code>.</p>
</div>
<div class="section level2">
<h2 id="load-data">2. Load data<a class="anchor" aria-label="anchor" href="#load-data"></a>
</h2>
<p>Here, we are using an exemplary mouse brain dataset from the <a href="https://info.vizgen.com/mouse-brain-map" class="external-link">Vizgen Data Release
V1.0</a>, Slice 2, Replicate 3, downloaded 08/2023. The dataset for this
tutorial can be downloaded <a href="https://drive.google.com/drive/folders/1hDILr2N2ZhNUutY1C2GNYrUX14E8nyDC?usp=sharing" class="external-link">here</a>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://themilolab.com/" class="external-link">SPATA2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SPATAData</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># initiate the object from a folder directory</span></span>
<span><span class="va">object</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="../reference/initiateSpataObjectMERFISH.html">initiateSpataObjectMERFISH</a></span><span class="op">(</span></span>
<span>     directory_merfish <span class="op">=</span> <span class="st">"data/tutorial_initiate_merfish"</span>, <span class="co"># adjust to your liking</span></span>
<span>     sample_name <span class="op">=</span> <span class="st">"Slice2_Replicate3"</span></span>
<span>   <span class="op">)</span></span>
<span></span>
<span><span class="co"># output object</span></span>
<span><span class="va">object</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class SPATA2</span></span>
<span><span class="co">## Sample: Slice2_Replicate3 </span></span>
<span><span class="co">## Size: 85958 x 649 (cells x molecules)</span></span>
<span><span class="co">## Memory: 179.16 Mb</span></span>
<span><span class="co">## Platform: MERFISH </span></span>
<span><span class="co">## Molecular assays (1):</span></span>
<span><span class="co">## 1. Assay</span></span>
<span><span class="co">##  Molecular modality: gene</span></span>
<span><span class="co">##  Distinct molecules: 649</span></span>
<span><span class="co">##  Matrices (1):</span></span>
<span><span class="co">##  -counts (active)</span></span>
<span><span class="co">## Meta variables (3): sample, original_barcodes, tissue_section</span></span></code></pre>
<p>We obtained a <code>SPATA2</code> object with 85,958 cells and 649
genes. The count matrix looks as follows:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># extract count matrix</span></span>
<span><span class="va">count_mtr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getCountMatrix.html">getCountMatrix</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># show results</span></span>
<span><span class="va">count_mtr</span><span class="op">[</span><span class="fl">10</span><span class="op">:</span><span class="fl">15</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## 6 x 5 sparse Matrix of class "dgCMatrix"</span></span>
<span><span class="co">##         cell_1 cell_2 cell_3 cell_4 cell_5</span></span>
<span><span class="co">## Htr5a        .      .      .      .      .</span></span>
<span><span class="co">## Htr5b        .      .      .      .      .</span></span>
<span><span class="co">## Htr6         .      .      .      .      .</span></span>
<span><span class="co">## Htr7         .      1      .      .      .</span></span>
<span><span class="co">## Adora1       .      .      .      .      .</span></span>
<span><span class="co">## Adora2a      .      .      .      .      .</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="spatial-processing">3. Spatial processing<a class="anchor" aria-label="anchor" href="#spatial-processing"></a>
</h2>
<p>First, we correct the slight tilt.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rotateAll.html">rotateCoordinates</a></span><span class="op">(</span><span class="va">object</span>, angle <span class="op">=</span> <span class="fl">6</span><span class="op">)</span></span></code></pre></div>
<p>Next, we plot the tissue outline. This has been computed via
<code><a href="../reference/identifyTissueOutline.html">identifyTissueOutline()</a></code> with default parameters within
<code><a href="../reference/initiateSpataObjectMERFISH.html">initiateSpataObjectMERFISH()</a></code>. If the results of your
objects are not satisfying adjust them and run the function again. Every
function call simply overwrites the results. In this case, the default
parameters worked out well and the tissue outline looks appropriate.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object</span>, color_by <span class="op">=</span> <span class="st">"tissue_section"</span>, pt_clrp <span class="op">=</span> <span class="st">"milo"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="../reference/ggpLayerTissueOutline.html">ggpLayerTissueOutline</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span></code></pre></div>
<p><img src="initiation-and-preprocessing-merfish_files/figure-html/unnamed-chunk-5-1.png" width="50%"></p>
<p>Seems as if there is a tiny fraction of cells that is not connected
to the main tissue section. We can identify and remove spatial outlier
separated from the contiguous tissue section with
<code><a href="../reference/identifySpatialOutliers.html">identifySpatialOutliers()</a></code>. Again, this function runs on
default parameters that can be adjusted. Refer to the documentation for
more information. In this case single outliers or collection of cells of
less than 5% of the overall number of cells (in this case, 4,297 cells)
are considered spatial outliers. Running the function creates a new
column <em>sp_outlier</em> which is added to the cell metadata.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># identify outliers based on the results of identifyTissueOutline()</span></span>
<span><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/identifySpatialOutliers.html">identifySpatialOutliers</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 02:00:39 Identifying spatial outliers.</span></span></code></pre>
<pre><code><span><span class="co">## 02:00:39 Spatial outliers: 95</span></span></code></pre>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># show the column</span></span>
<span><span class="fu"><a href="../reference/getCoordsDf.html">getCoordsDf</a></span><span class="op">(</span><span class="va">object</span>, variables <span class="op">=</span> <span class="st">"sp_outlier"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">sp_outlier</span> <span class="op">==</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">barcodes</span>, <span class="va">sample</span>, <span class="va">sp_outlier</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/everything.html" class="external-link">everything</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 95 Ã— 13</span></span></span>
<span><span class="co">##    barcodes sample sp_outlier     x     y x_orig y_orig   fov volume min_x max_x</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> cell_16â€¦ Sliceâ€¦ TRUE       <span style="text-decoration: underline;">1</span>407. <span style="text-decoration: underline;">6</span>327.  <span style="text-decoration: underline;">1</span>407.  <span style="text-decoration: underline;">6</span>327.    98  <span style="text-decoration: underline;">1</span>017. <span style="text-decoration: underline;">1</span>139. <span style="text-decoration: underline;">1</span>165.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> cell_28â€¦ Sliceâ€¦ TRUE        727. <span style="text-decoration: underline;">1</span>085.   727.  <span style="text-decoration: underline;">1</span>085.   124  <span style="text-decoration: underline;">1</span>460. <span style="text-decoration: underline;">1</span>010. <span style="text-decoration: underline;">1</span>038.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> cell_33â€¦ Sliceâ€¦ TRUE       <span style="text-decoration: underline;">4</span>002.  499.  <span style="text-decoration: underline;">4</span>002.   499.   648   297. <span style="text-decoration: underline;">4</span>335. <span style="text-decoration: underline;">4</span>349.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> cell_33â€¦ Sliceâ€¦ TRUE       <span style="text-decoration: underline;">4</span>033.  508.  <span style="text-decoration: underline;">4</span>033.   508.   648   749. <span style="text-decoration: underline;">4</span>364. <span style="text-decoration: underline;">4</span>381.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> cell_33â€¦ Sliceâ€¦ TRUE       <span style="text-decoration: underline;">4</span>009.  533.  <span style="text-decoration: underline;">4</span>009.   533.   648  <span style="text-decoration: underline;">1</span>717. <span style="text-decoration: underline;">4</span>336. <span style="text-decoration: underline;">4</span>355.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> cell_33â€¦ Sliceâ€¦ TRUE       <span style="text-decoration: underline;">4</span>035.  435.  <span style="text-decoration: underline;">4</span>035.   435.   649  <span style="text-decoration: underline;">1</span>650. <span style="text-decoration: underline;">4</span>375. <span style="text-decoration: underline;">4</span>389.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> cell_33â€¦ Sliceâ€¦ TRUE       <span style="text-decoration: underline;">4</span>025.  481.  <span style="text-decoration: underline;">4</span>025.   481.   649   439. <span style="text-decoration: underline;">4</span>362. <span style="text-decoration: underline;">4</span>372.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> cell_33â€¦ Sliceâ€¦ TRUE       <span style="text-decoration: underline;">3</span>982.  451.  <span style="text-decoration: underline;">3</span>982.   451.   649   180. <span style="text-decoration: underline;">4</span>322. <span style="text-decoration: underline;">4</span>332.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> cell_33â€¦ Sliceâ€¦ TRUE       <span style="text-decoration: underline;">4</span>001.  459.  <span style="text-decoration: underline;">4</span>001.   459.   649   720. <span style="text-decoration: underline;">4</span>340. <span style="text-decoration: underline;">4</span>351.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> cell_33â€¦ Sliceâ€¦ TRUE       <span style="text-decoration: underline;">4</span>006.  427.  <span style="text-decoration: underline;">4</span>006.   427.   649   503. <span style="text-decoration: underline;">4</span>350. <span style="text-decoration: underline;">4</span>357.</span></span>
<span><span class="co">## <span style="color: #949494;"># â„¹ 85 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># â„¹ 2 more variables: min_y &lt;dbl&gt;, max_y &lt;dbl&gt;</span></span></span></code></pre>
<p>The results can be plotted with <code><a href="../reference/plotSurface.html">plotSurface()</a></code>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># before</span></span>
<span><span class="va">outline_with_outliers</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ggpLayerTissueOutline.html">ggpLayerTissueOutline</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot_with_outliers</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object</span>, color_by <span class="op">=</span> <span class="st">"sp_outlier"</span>, pt_clrp <span class="op">=</span> <span class="st">"milo"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># remove the outliers</span></span>
<span><span class="co"># note that with spatial_proc = TRUE (default), a new tissue outline is computed</span></span>
<span><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/removeSpatialOutliers.html">removeSpatialOutliers</a></span><span class="op">(</span><span class="va">object</span>, spatial_proc <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># afterwards</span></span>
<span><span class="va">plot_without_outliers</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object</span>, color_by <span class="op">=</span> <span class="st">"sp_outlier"</span>, pt_clrp <span class="op">=</span> <span class="st">"milo"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># left plot</span></span>
<span><span class="va">plot_with_outliers</span> <span class="op">+</span> </span>
<span>  <span class="va">outline_with_outliers</span> <span class="co"># old outline</span></span>
<span></span>
<span><span class="co"># right plot</span></span>
<span><span class="va">plot_without_outliers</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="../reference/ggpLayerTissueOutline.html">ggpLayerTissueOutline</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span> <span class="co"># new outline</span></span></code></pre></div>
<p><img src="initiation-and-preprocessing-merfish_files/figure-html/unnamed-chunk-7-1.png" width="50%"><img src="initiation-and-preprocessing-merfish_files/figure-html/unnamed-chunk-7-2.png" width="50%"></p>
</div>
<div class="section level2">
<h2 id="data-preprocessing">4. Data preprocessing<a class="anchor" aria-label="anchor" href="#data-preprocessing"></a>
</h2>
<p>The first processing step is usually to remove genes that were not
detected in any cell. As well as to remove cells with no counts at all.
The functions below check if any such scenario is the case and if so,
they remove what needs to be removed. Else, they simply return the input
object. In either case they give feedback as long
<code>verbose = TRUE</code> (the default).</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">#--- gene removal </span></span>
<span><span class="fu"><a href="../reference/nMolecules.html">nGenes</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span><span class="co">## [1] 649</span></span>
<span></span>
<span><span class="co"># remove genes with no counts - none existing</span></span>
<span><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/removeMolecules.html">removeGenesZeroCounts</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/nMolecules.html">nGenes</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span><span class="co">## [1] 649</span></span>
<span></span>
<span><span class="co">#--- obs removal</span></span>
<span><span class="fu"><a href="../reference/nObs.html">nObs</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span><span class="co">## [1] 85863</span></span>
<span></span>
<span><span class="co"># remove observations with no counts - none existing</span></span>
<span><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/removeObs.html">removeObsZeroCounts</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/nObs.html">nObs</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span><span class="co">## [1] 85863</span></span></code></pre></div>
<p>Next, we normalize the count matrix. We normalize using SCT as
described in the <a href="https://satijalab.org/seurat/articles/seurat5_spatial_vignette_2#mouse-brain-vizgen-merscope" class="external-link">Seurat
utorial</a>. This processed the raw count matrix. Results are stored in
the respective assays. By default, the normalized matrix is activated
and thus used in downstream analysis. See <code><a href="../reference/activateMatrix.html">?activateMatrix</a></code>
for more information. Furthermore, we compute metadata for the
cells:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># no processed matrices, only raw counts</span></span>
<span><span class="fu"><a href="../reference/getProcessedMatrixNames.html">getProcessedMatrixNames</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span><span class="co">## character(0)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/getMatrixNames.html">getMatrixNames</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span><span class="co">## [1] "counts"</span></span>
<span></span>
<span><span class="fu"><a href="../reference/activateMatrix.html">activeMatrix</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span><span class="co">## [1] "counts"</span></span>
<span></span>
<span><span class="co"># normalize raw count data</span></span>
<span><span class="va">object</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/normalizeCounts.html">normalizeCounts</a></span><span class="op">(</span></span>
<span>    object <span class="op">=</span> <span class="va">object</span>,</span>
<span>    method <span class="op">=</span> <span class="st">"SCT"</span>,</span>
<span>    mtr_name_new <span class="op">=</span> <span class="st">"SCT_data"</span>, </span>
<span>    sct_clip_range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">10</span>, <span class="fl">10</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="co"># now the object contains normalized data</span></span>
<span><span class="fu"><a href="../reference/getProcessedMatrixNames.html">getProcessedMatrixNames</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span><span class="co">## [1] "SCT_data"</span></span>
<span></span>
<span><span class="fu"><a href="../reference/activateMatrix.html">activeMatrix</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span><span class="co">## [1] "SCT_data"</span></span></code></pre></div>
<p>By default, the normalized matrix is activated and thus used in
downstream analysis. See <code><a href="../reference/activateMatrix.html">?activateMatrix</a></code> for more
information. Furthermore, we compute metadata for the cells:
<em>n_counts_{modality}</em> with molecular modality <em>gene</em>
represents the number of individual transcript counts per cell, whereas
<em>n_distinct_{modality}</em> represents the number of distinct genes
identified by cell.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># some more meta features</span></span>
<span><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/computeMetaFeatures.html">computeMetaFeatures</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># meta feauture names</span></span>
<span><span class="fu"><a href="../reference/getMetaNames.html">getFeatureNames</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span><span class="co">##           character             logical              factor             numeric </span></span>
<span><span class="co">## "original_barcodes"        "sp_outlier"    "tissue_section"     "n_counts_gene" </span></span>
<span><span class="co">##             integer             numeric </span></span>
<span><span class="co">##   "n_distinct_gene"      "avg_cpm_gene"</span></span>
<span></span>
<span><span class="co"># show overview (compare to overview above)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/methods/show.html" class="external-link">show</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span><span class="co">## An object of class SPATA2</span></span>
<span><span class="co">## Sample: Slice2_Replicate3 </span></span>
<span><span class="co">## Size: 85863 x 649 (cells x molecules)</span></span>
<span><span class="co">## Memory: 261.64 Mb</span></span>
<span><span class="co">## Platform: MERFISH </span></span>
<span><span class="co">## Molecular assays (1):</span></span>
<span><span class="co">## 1. Assay</span></span>
<span><span class="co">##  Molecular modality: gene</span></span>
<span><span class="co">##  Distinct molecules: 649</span></span>
<span><span class="co">##  Matrices (2):</span></span>
<span><span class="co">##  -counts</span></span>
<span><span class="co">##  -SCT_data (active)</span></span>
<span><span class="co">## Meta variables (7): sample, original_barcodes, sp_outlier, tissue_section, n_counts_gene, n_distinct_gene, avg_cpm_gene</span></span>
<span></span>
<span><span class="co"># left plot</span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object</span>, color_by <span class="op">=</span> <span class="st">"n_counts_gene"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># right plot</span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object</span>, color_by <span class="op">=</span> <span class="st">"n_distinct_gene"</span><span class="op">)</span></span></code></pre></div>
<p><img src="initiation-and-preprocessing-merfish_files/figure-html/unnamed-chunk-10-1.png" width="50%"><img src="initiation-and-preprocessing-merfish_files/figure-html/unnamed-chunk-10-2.png" width="50%"></p>
</div>
<div class="section level2">
<h2 id="spatially-variable-genes">5. Spatially variable genes<a class="anchor" aria-label="anchor" href="#spatially-variable-genes"></a>
</h2>
<p>Next, we identify spatially variable genes using SPARK-X.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># run the algorithm</span></span>
<span><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runSPARKX.html">runSPARKX</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span><span class="co">## ## ===== SPARK-X INPUT INFORMATION ==== </span></span>
<span><span class="co">## ## number of total samples: 85863 </span></span>
<span><span class="co">## ## number of total genes: 649 </span></span>
<span><span class="co">## ## Running with single core, may take some time </span></span>
<span><span class="co">## ## Testing With Projection Kernel</span></span>
<span><span class="co">## ## Testing With Gaussian Kernel 1</span></span>
<span><span class="co">## ## Testing With Gaussian Kernel 2</span></span>
<span><span class="co">## ## Testing With Gaussian Kernel 3</span></span>
<span><span class="co">## ## Testing With Gaussian Kernel 4</span></span>
<span><span class="co">## ## Testing With Gaussian Kernel 5</span></span>
<span><span class="co">## ## Testing With Cosine Kernel 1</span></span>
<span><span class="co">## ## Testing With Cosine Kernel 2</span></span>
<span><span class="co">## ## Testing With Cosine Kernel 3</span></span>
<span><span class="co">## ## Testing With Cosine Kernel 4</span></span>
<span><span class="co">## ## Testing With Cosine Kernel 5</span></span>
<span></span>
<span><span class="va">sparkx_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runSPARKX.html">getSparkxGenes</a></span><span class="op">(</span><span class="va">object</span>, threshold_pval <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">sparkx_genes</span><span class="op">)</span></span>
<span><span class="co">##  chr [1:570] "Htr1f" "Htr2a" "Htr5a" "Htr5b" "Adora1" "Adora2a" "Adgrb1" ...</span></span>
<span></span>
<span><span class="co"># left plot</span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object</span>, color_by <span class="op">=</span> <span class="va">sparkx_genes</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, pt_clrsp <span class="op">=</span> <span class="st">"Purple-Blue"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="../reference/ggpLayerTissueOutline.html">ggpLayerTissueOutline</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span> <span class="co"># use outline to plot against white</span></span>
<span></span>
<span><span class="co"># right plot (set the color scale limits)</span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object</span>, color_by <span class="op">=</span> <span class="va">sparkx_genes</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, pt_clrsp <span class="op">=</span> <span class="st">"Reds 3"</span>, limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1.5</span><span class="op">)</span>, oob <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/oob.html" class="external-link">squish</a></span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="../reference/ggpLayerTissueOutline.html">ggpLayerTissueOutline</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span></code></pre></div>
<p><img src="initiation-and-preprocessing-merfish_files/figure-html/unnamed-chunk-11-1.png" width="50%"><img src="initiation-and-preprocessing-merfish_files/figure-html/unnamed-chunk-11-2.png" width="50%"></p>
</div>
<div class="section level2">
<h2 id="subset-observations">6. Subset observations<a class="anchor" aria-label="anchor" href="#subset-observations"></a>
</h2>
<p>For examples, we provide a small example object. Here, we
additionally subset the dataset to just extract the top right
corner.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># get range of upper right corner</span></span>
<span><span class="va">crop_range</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/getCoordsRange.html">getCoordsRange</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map</a></span><span class="op">(</span>.f <span class="op">=</span> <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">.x</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">.x</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># show results</span></span>
<span><span class="va">crop_range</span></span>
<span><span class="co">## $x</span></span>
<span><span class="co">## [1] 5069.762 9639.579</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $y</span></span>
<span><span class="co">## [1] 3732.895 7104.322</span></span>
<span></span>
<span><span class="va">p_before</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object</span>, pt_clr <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="../reference/ggpLayerTissueOutline.html">ggpLayerTissueOutline</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="../reference/ggpLayerRect.html">ggpLayerRect</a></span><span class="op">(</span><span class="va">object</span>, xrange <span class="op">=</span> <span class="va">crop_range</span><span class="op">$</span><span class="va">x</span>, yrange <span class="op">=</span> <span class="va">crop_range</span><span class="op">$</span><span class="va">y</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># crop with a rectangular</span></span>
<span><span class="va">object_subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cropSpataObject.html">cropSpataObject</a></span><span class="op">(</span><span class="va">object</span>, xrange <span class="op">=</span> <span class="va">crop_range</span><span class="op">$</span><span class="va">x</span>, yrange <span class="op">=</span> <span class="va">crop_range</span><span class="op">$</span><span class="va">y</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p_subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object_subset</span>, pt_clr <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># left plot</span></span>
<span><span class="va">p_before</span></span>
<span></span>
<span><span class="co"># right plot</span></span>
<span><span class="va">p_subset</span></span></code></pre></div>
<p><img src="initiation-and-preprocessing-merfish_files/figure-html/unnamed-chunk-12-1.png" width="50%"><img src="initiation-and-preprocessing-merfish_files/figure-html/unnamed-chunk-12-2.png" width="50%"></p>
</div>
<div class="section level2">
<h2 id="conclusion-and-more-data-sets">7. Conclusion and more data sets<a class="anchor" aria-label="anchor" href="#conclusion-and-more-data-sets"></a>
</h2>
<p>Thatâ€™s it. The object can be used for any downstream analyses such as
<a href="dimensional-reduction.html">dimensional reduction</a>, <a href="clustering.html">clustering</a>, <a href="spatial-annotation-screening.html">spatial annotation
screening</a> or <a href="spatial-trajectory-screening.html">spatial
trajectory screening</a>. Refer to tab Tutorials for more links.
Furthermore, you can skim our curated data base of spatial data sets for
those of platform MERFISH using <a href="spata-data.html">SPATAData</a>.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load package</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SPATAData</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># filter for samples from platform VisiumHD</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SPATAData/man/sourceDataFrame.html" class="external-link">sourceDataFrame</a></span><span class="op">(</span><span class="va">platform</span> <span class="op">==</span> <span class="st">"MERFISH"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 32 Ã— 20</span></span></span>
<span><span class="co">##    sample_name    donor_species institution lm_source           organ organ_part</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dttm&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> HumanBreastCaâ€¦ Homo sapiens  Vizgen      2024-08-24 <span style="color: #949494;">01:57:29</span> Breaâ€¦ <span style="color: #BB0000;">NA</span>        </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> HumanColonCanâ€¦ Homo sapiens  Vizgen      2024-08-24 <span style="color: #949494;">01:57:29</span> Colon <span style="color: #BB0000;">NA</span>        </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> HumanColonCanâ€¦ Homo sapiens  Vizgen      2024-08-24 <span style="color: #949494;">01:57:29</span> Colon <span style="color: #BB0000;">NA</span>        </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> HumanLiverCanâ€¦ Homo sapiens  Vizgen      2024-08-24 <span style="color: #949494;">01:57:29</span> Liver <span style="color: #BB0000;">NA</span>        </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> HumanLiverCanâ€¦ Homo sapiens  Vizgen      2024-08-24 <span style="color: #949494;">01:57:29</span> Liver <span style="color: #BB0000;">NA</span>        </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> HumanLungCancâ€¦ Homo sapiens  Vizgen      2024-08-24 <span style="color: #949494;">01:57:29</span> Lung  <span style="color: #BB0000;">NA</span>        </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> HumanLungCancâ€¦ Homo sapiens  Vizgen      2024-08-24 <span style="color: #949494;">01:57:29</span> Lung  <span style="color: #BB0000;">NA</span>        </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> HumanMelanomaâ€¦ Homo sapiens  Vizgen      2024-08-24 <span style="color: #949494;">01:57:29</span> Skin  <span style="color: #BB0000;">NA</span>        </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> HumanMelanomaâ€¦ Homo sapiens  Vizgen      2024-08-24 <span style="color: #949494;">01:57:29</span> Skin  <span style="color: #BB0000;">NA</span>        </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> HumanOvarianCâ€¦ Homo sapiens  Vizgen      2024-08-24 <span style="color: #949494;">01:57:29</span> Ovary <span style="color: #BB0000;">NA</span>        </span></span>
<span><span class="co">## <span style="color: #949494;"># â„¹ 22 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># â„¹ 14 more variables: pathology &lt;chr&gt;, platform &lt;chr&gt;, pub_citation &lt;chr&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   source &lt;chr&gt;, web_link &lt;chr&gt;, mean_counts &lt;dbl&gt;, median_counts &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   modality_gene &lt;lgl&gt;, modality_metabolite &lt;lgl&gt;, modality_protein &lt;lgl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   n_obs &lt;int&gt;, n_tissue_sections &lt;int&gt;, obs_unit &lt;chr&gt;, obj_size &lt;lbstr_by&gt;</span></span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Jan Kueckelhaus, Dieter-Henrik Heiland.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
