<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Molecular Variables • SPATA2</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Molecular Variables">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SPATA2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">3.1.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li>
  <a href="../articles/spata-data.html">SPATAData</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Initiating a SPATA2 object</li>
    <li>
      <a href="../articles/initiation-and-preprocessing-customized.html">Initiation &amp; Preprocessing with individual data</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-merfish.html">Initiation &amp; Preprocessing with MERFISH data</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-visium.html">Initiation &amp; Preprocessing with Visium data</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-visium-hd.html">Initiation &amp; Preprocessing with VisiumHD data</a>
    </li>
    <li class="dropdown-header">Basic Manipulation of a SPATA2 object</li>
    <li>
      <a href="../articles/adding-data.html">Adding Data to SPATA2 objects</a>
    </li>
    <li>
      <a href="../articles/package-compatibility.html">Platform Compatibility</a>
    </li>
    <li>
      <a href="../articles/molecular-variables.html">Molecular Variables</a>
    </li>
    <li>
      <a href="../articles/image-handling.html">Image Handling</a>
    </li>
    <li>
      <a href="../articles/spatial-segmentation.html">Spatial Segmentation</a>
    </li>
    <li class="dropdown-header">Implemented Algorithms &amp; Concepts</li>
    <li>
      <a href="../articles/cnv.html">Copy Number Variations</a>
    </li>
    <li>
      <a href="../articles/clustering.html">Clustering</a>
    </li>
    <li>
      <a href="../articles/dimensional-reduction.html">Dimensional Reduction</a>
    </li>
    <li>
      <a href="../articles/dea.html">Differential Expression</a>
    </li>
    <li>
      <a href="../articles/gsea.html">Gene Set Enrichment</a>
    </li>
    <li>
      <a href="../articles/spatial-measures.html">Spatial Measures</a>
    </li>
    <li class="dropdown-header">Spatial Gradient Screening</li>
    <li>
      <a href="../articles/creating-spatial-annotations.html">Creating Spatial Annotations</a>
    </li>
    <li>
      <a href="../articles/using-spatial-annotations.html">Using Spatial Annotations</a>
    </li>
    <li>
      <a href="../articles/spatial-annotation-screening.html">Spatial Annotation Screening</a>
    </li>
    <li>
      <a href="../articles/creating-spatial-trajectories.html">Creating Spatial Trajectories</a>
    </li>
    <li>
      <a href="../articles/spatial-trajectory-screening.html">Spatial Trajectory Screening</a>
    </li>
    <li class="dropdown-header">Visualization</li>
    <li>
      <a href="../articles/plotting-surface.html">Surface Plotting</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/newslog.html">Newslog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Molecular Variables</h1>
            
      

      <div class="hidden name"><code>molecular-variables.Rmd</code></div>

    </div>

    
    
<div class="section level3">
<h3 id="introduction">1. Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load required packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://themilolab.com/" class="external-link">SPATA2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># load SPATA2 inbuilt example data</span></span>
<span><span class="va">object_t269</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadExampleObject.html">loadExampleObject</a></span><span class="op">(</span><span class="st">"UKF269T"</span>, process <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/methods/show.html" class="external-link">show</a></span><span class="op">(</span><span class="va">object_t269</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/methods/show.html" class="external-link">show</a></span><span class="op">(</span><span class="va">object_t269</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class SPATA2</span></span>
<span><span class="co">## Sample: UKF269T_diet </span></span>
<span><span class="co">## Size: 3213 x 15000 (spots x molecules)</span></span>
<span><span class="co">## Memory: 192.59 Mb</span></span>
<span><span class="co">## Platform: VisiumSmall </span></span>
<span><span class="co">## Molecular assays (1):</span></span>
<span><span class="co">## 1. Assay</span></span>
<span><span class="co">##  Molecular modality: gene</span></span>
<span><span class="co">##  Distinct molecules: 15000</span></span>
<span><span class="co">##  Matrices (2):</span></span>
<span><span class="co">##  -counts</span></span>
<span><span class="co">##  -LogNormalize (active)</span></span>
<span><span class="co">## Registered images (1):</span></span>
<span><span class="co">## - lowres (582x600 px, active, loaded)</span></span>
<span><span class="co">## Meta variables (6): sample, sp_outlier, tissue_section, n_counts_gene, n_distinct_gene, avg_cpm_gene</span></span>
<span><span class="co">## Spatial Trajectories: horizontal_mid</span></span></code></pre>
<p>Molecular variables are those that are directly linked to molecular
expression data of the <code>SPATA2</code> object, like gene expression
or gene sets. The concept of genes and gene sets has been embedded in a
broader framework of molecular data since spatial multi-omic studies do
not only include gene expression but also proteins or metabolites.
Developed initially for the Visium platform, which focuses on spatial
gene expression, SPATA2 aims to evolve to support the analysis of
various molecular data types. Hence, we introduce the term molecular
modalities to refer to the different types of molecular data that can be
analyzed within the SPATA2 package.</p>
<p>In SPATA2, the concept of molecular modalities is crucial as it
ensures that specific functions and analyses are applied correctly based
on the type of molecular data. While you can always specifiy your own
modality, the package recognizes three primary molecular modalities:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Gene expression</strong>: Represented by
<code>modality = 'gene'</code>. This modality deals with the spatial
distribution and analysis of mRNA transcripts.</li>
<li>
<strong>Protein expression</strong>: Represented by
<code>modality = 'protein'</code>. This modality focuses on the spatial
distribution and analysis of protein expression data.</li>
<li>
<strong>Metabolites expression</strong>: Represented by
<code>modality = 'metabolite'</code>. This modality is concerned with
the spatial distribution and analysis of metabolite data.</li>
</ol>
<p>These modalities are essential for performing specific functions as
well as dealing with molecular signatures - sets of molecules (such as
genes or proteins) that are associated with specific biological states,
processes, or conditions. This tutorial guides you through how to work
with SPATA2 functions that adress this concept.</p>
</div>
<div class="section level2">
<h2 id="molecule-names">2. Molecule names<a class="anchor" aria-label="anchor" href="#molecule-names"></a>
</h2>
<p>Whether you are analyzing transcriptomic, proteomic or metabolomic
data your assays will feature counts of molecules (mRNA transcripts,
proteins or metabolites). The names of these molecules that are stored
in the molecular assays of your <code>SPATA2</code> object can be
extracted as character vectors using either <code><a href="../reference/getMolecules.html">getMolecules()</a></code>
with the argument <code>assay_name</code> specified (defaults to the
active assay, not relevant if there exists only one assay in the object)
or with “wrapper-functions” <code><a href="../reference/getMolecules.html">getGenes()</a></code>,
<code><a href="../reference/getMolecules.html">getProteins()</a></code> and <code><a href="../reference/getMolecules.html">getMetabolites()</a></code> depending
on the molecular modality of the assay(s) in your <code>SPATA2</code>
object.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># contains only one assay (molecular modality: gene)</span></span>
<span><span class="fu"><a href="../reference/getAssayModalities.html">getAssayNames</a></span><span class="op">(</span><span class="va">object_t269</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "gene"</span></span></code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/activateAssay.html">activeAssay</a></span><span class="op">(</span><span class="va">object_t269</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "gene"</span></span></code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># use the wrapper function</span></span>
<span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getMolecules.html">getGenes</a></span><span class="op">(</span><span class="va">object_t269</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">genes</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  chr [1:15000] "AL669831.5" "FAM87B" "FAM41C" "AL645608.1" "KLHL17" "PERM1" ...</span></span></code></pre>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># getMolecules() works, too</span></span>
<span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getMolecules.html">getMolecules</a></span><span class="op">(</span><span class="va">object_t269</span>, assay_name <span class="op">=</span> <span class="st">"gene"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">genes</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  chr [1:15000] "AL669831.5" "FAM87B" "FAM41C" "AL645608.1" "KLHL17" "PERM1" ...</span></span></code></pre>
<p>Using <code><a href="../reference/getMolecules.html">getProteins()</a></code> on a <code>SPATA2</code> object that
does not contain an assay of molecular modality <em>protein</em> will
fail.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/containsModality.html">containsModality</a></span><span class="op">(</span><span class="va">object_t269</span>, modality <span class="op">=</span> <span class="st">"protein"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] FALSE</span></span></code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/conditions.html" class="external-link">tryCatch</a></span><span class="op">(</span></span>
<span>  expr <span class="op">=</span> <span class="op">{</span> <span class="fu"><a href="../reference/getMolecules.html">getProteins</a></span><span class="op">(</span><span class="va">object_t269</span><span class="op">)</span> <span class="op">}</span>,</span>
<span>  error <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">error</span><span class="op">)</span><span class="op">{</span> <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span><span class="op">(</span><span class="va">error</span><span class="op">$</span><span class="va">message</span><span class="op">)</span><span class="op">}</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## error in evaluating the argument 'object' in selecting a method for function 'getCountMatrix': This SPATA2 object does not contain an assay of molecular modality 'protein'.</span></span></code></pre>
<p>Since we only have gene expression data stored in the example object,
we can stick to the <code><a href="../reference/getMolecules.html">getGenes()</a></code> wrapper. Note, that you can
not only use this function to extract the rownames of the count matrix
of the respective assay. You can also use it to subset the genes by
molecular signatures (see section on molecular signatures for more
information).</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## using signatures argument</span></span>
<span><span class="co"># character vector of molecules of specific signatures (simplify defaults to TRUE)</span></span>
<span><span class="va">synapse_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getMolecules.html">getGenes</a></span><span class="op">(</span><span class="va">object_t269</span>, signatures <span class="op">=</span> <span class="st">"CC.GO_SYNAPSE"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">synapse_genes</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  chr [1:744] "CDH2" "GJC1" "CDH8" "CDH9" "CDH10" "GPC6" "USH1C" "CDH11" ...</span></span></code></pre>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSurfaceComparison.html">plotSurfaceComparison</a></span><span class="op">(</span><span class="va">object_t269</span>, color_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="va">synapse_genes</span>, <span class="fl">9</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="molecular-variables_files/figure-html/unnamed-chunk-5-1.png" width="100%"></p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># list of gene sets</span></span>
<span><span class="va">tcr_gs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"RCTM_TCR_SIGNALING"</span>, <span class="st">"RCTM_DOWNSTREAM_TCR_SIGNALING"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># simplify = TRUE (default) merges the output to a character vector of unique names</span></span>
<span><span class="va">tcr_genes_vec</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getMolecules.html">getGenes</a></span><span class="op">(</span><span class="va">object_t269</span>, signatures <span class="op">=</span> <span class="va">tcr_gs</span>, simplify <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">tcr_genes_vec</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  chr [1:101] "PSMB1" "CD4" "WAS" "PSMA4" "LCP2" "TAB2" "UBE2D1" "PAG1" ...</span></span></code></pre>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># simplify = FALSE returns a list</span></span>
<span><span class="va">tcr_genes_lst</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getMolecules.html">getGenes</a></span><span class="op">(</span><span class="va">object_t269</span>, signatures <span class="op">=</span> <span class="va">tcr_gs</span>, simplify <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">tcr_genes_lst</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## List of 2</span></span>
<span><span class="co">##  $ RCTM_TCR_SIGNALING           : chr [1:58] "PSMB1" "CD4" "WAS" "PSMA4" ...</span></span>
<span><span class="co">##  $ RCTM_DOWNSTREAM_TCR_SIGNALING: chr [1:43] "PSMB1" "CD4" "PSMA4" "TAB2" ...</span></span></code></pre>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">tcr_genes_lst</span>, <span class="va">head</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $RCTM_TCR_SIGNALING</span></span>
<span><span class="co">## [1] "PSMB1" "CD4"   "WAS"   "PSMA4" "LCP2"  "TAB2" </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $RCTM_DOWNSTREAM_TCR_SIGNALING</span></span>
<span><span class="co">## [1] "PSMB1"  "CD4"    "PSMA4"  "TAB2"   "UBE2D1" "PSME1"</span></span></code></pre>
<p>The same concept applies to members of the
<code><a href="../reference/getMolecules.html">getMolecules()</a></code> family of functions:
<code><a href="../reference/getMolecules.html">getMetabolites()</a></code> and <code><a href="../reference/getMolecules.html">getProteins()</a></code>.</p>
</div>
<div class="section level2">
<h2 id="molecular-signatures">3. Molecular signatures<a class="anchor" aria-label="anchor" href="#molecular-signatures"></a>
</h2>
<p>In SPATA2, a molecular signature is represented as a vector in a
named list, where the character values are the molecules of which the
signature consists. Included in the package is a list named
<code>signatures</code>, with corresponding slots
<code>signatures$gene</code>, <code>signatures$protein</code>, and
<code>signatures$metabolite</code>. This list is where default
signatures are stored for the respective data modality. Depending on how
the <code>SPATA2</code> object is initiated, the created molecular
assays already contain the respective signatures in the
<code>signatures</code> slot. The scores of the signatures are computed
upon extraction or upon usage for computation and visualization using
the active data matrix of the respective assay.</p>
<div class="section level3">
<h3 id="extract-molecular-signature-names">3.1 Extract molecular signature names<a class="anchor" aria-label="anchor" href="#extract-molecular-signature-names"></a>
</h3>
<p>To obtain the names of the signatures that are currently stored, use
<code><a href="../reference/getSignatureNames.html">getSignatureNames()</a></code> or it’s wrappers:
<code><a href="../reference/getSignatureNames.html">getGeneSets()</a></code>, <code><a href="../reference/getSignatureNames.html">getProteinSets()</a></code> and
<code><a href="../reference/getSignatureNames.html">getMetaboliteSets()</a></code> depending on the molecular modality of
interest.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## extraction</span></span>
<span><span class="co"># opt 1</span></span>
<span><span class="va">all_signatures</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getSignatureNames.html">getSignatureNames</a></span><span class="op">(</span><span class="va">object_t269</span>, assay_name <span class="op">=</span> <span class="st">"gene"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">all_signatures</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  chr [1:11654] "BC_41BB_PATHWAY" "BC_ACE2_PATHWAY" ...</span></span></code></pre>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># whether you specify assay_name or not does not make a difference since</span></span>
<span><span class="co"># the object only contains one assay</span></span>
<span><span class="va">hallmark_signatures</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getSignatureNames.html">getSignatureNames</a></span><span class="op">(</span><span class="va">object_t269</span>, class <span class="op">=</span> <span class="st">"HM"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">hallmark_signatures</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  chr [1:50] "HM_ADIPOGENESIS" "HM_ALLOGRAFT_REJECTION" ...</span></span></code></pre>
<p>The code chunk above exemplifies how you can subset the extracted
signatures by class. In SPATA2, a signature name consists of two
parts:</p>
<p><em>class</em> _ <em>biological function</em></p>
<p>For instance, the gene set <em>HM_HYPOXIA</em> is of class
<em>HM</em> (short for Hallmark) and contains genes associated with
increased presence of or response to hypoxic circumstances. The class
indicates the source from where the signature derives and is separated
from the biological function part with the <strong>first</strong>
<code>_</code>. Underscores afterwards are ignored and interpreted as
part of the biological function, as in <em>RCTM_TCR_SIGNALING</em>
(class: <em>RCTM</em>; biological function: <em>TCR_SIGNALING</em>).</p>
</div>
<div class="section level3">
<h3 id="add-molecular-signatures">3.2 Add molecular signatures<a class="anchor" aria-label="anchor" href="#add-molecular-signatures"></a>
</h3>
<p>To add a molecular signature, or gene set, to a <code>SPATA2</code>
object (for example, the <code>UKF269T</code> object with its gene
expression assay), you can use the <code><a href="../reference/addSignature.html">addSignature()</a></code> or its
modality specific functions like <code><a href="../reference/addSignature.html">addGeneSet()</a></code>. The
<code>class</code> argument defaults to <em>‘UD’</em> (short for user
defined), indicating that the signature was added by the user. While in
this example the <code>class</code> argument is explicitly specified, it
is not necessary to do so, since it is the default. You can provide
different input for this argument if you want to categorize your
individually added molecular signatures into different classes for
easier extraction and organization.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># example genes of an hypothetical gene signature (gene set)</span></span>
<span><span class="va">genes</span> <span class="op">&lt;-</span></span>
<span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'CYFIP1'</span>, <span class="st">'SLC16A3'</span>, <span class="st">'AKAP5'</span>, <span class="st">'ADCY8'</span>, <span class="st">'CALB2'</span>, <span class="st">'GRIN1'</span>, <span class="st">'NLGN4X'</span>, <span class="st">'NLGN1'</span>,</span>
<span>   <span class="st">'ITGA3'</span>, <span class="st">'NLGN4Y'</span>, <span class="st">'ELFN1'</span>, <span class="st">'BSN'</span>, <span class="st">'CNTN6'</span>, <span class="st">'PDE4B'</span>, <span class="st">'DGKI'</span>, <span class="st">'LRRTM2'</span>, <span class="st">'LRRTM1'</span>,</span>
<span>   <span class="st">'SRPX2'</span>, <span class="st">'SHANK1'</span>, <span class="st">'SLC17A7'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># both, opt1 and opt2, have the same effect (addSignature() just allows to specify the</span></span>
<span><span class="co"># assay of interest, which is fixed to 'gene' for addGeneSet()).</span></span>
<span><span class="co"># opt1</span></span>
<span><span class="va">object_t269</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/addSignature.html">addSignature</a></span><span class="op">(</span><span class="va">object_t269</span>, class <span class="op">=</span> <span class="st">"UD"</span>, name <span class="op">=</span> <span class="st">"EXCITATORY_SYNAPSE"</span>, molecules <span class="op">=</span> <span class="va">genes</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># opt2</span></span>
<span><span class="co"># object &lt;- addGeneSet(object, name = "EXCITATORY_SYNAPSE", class = "HM", genes = genes)</span></span>
<span></span>
<span><span class="va">gs</span> <span class="op">&lt;-</span> <span class="st">"UD_EXCITATORY_SYNAPSE"</span></span>
<span></span>
<span><span class="co"># extract genes of signature</span></span>
<span><span class="fu"><a href="../reference/getMolecules.html">getGenes</a></span><span class="op">(</span><span class="va">object_t269</span>, signatures <span class="op">=</span> <span class="va">gs</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "CYFIP1"  "SLC16A3" "AKAP5"   "ADCY8"   "CALB2"   "GRIN1"   "NLGN4X" </span></span>
<span><span class="co">##  [8] "NLGN1"   "ITGA3"   "NLGN4Y"  "ELFN1"   "BSN"     "CNTN6"   "PDE4B"  </span></span>
<span><span class="co">## [15] "DGKI"    "LRRTM2"  "LRRTM1"  "SRPX2"   "SHANK1"  "SLC17A7"</span></span></code></pre>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># visualize signature expression</span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object_t269</span>, color_by <span class="op">=</span> <span class="va">gs</span><span class="op">)</span></span></code></pre></div>
<p><img src="molecular-variables_files/figure-html/unnamed-chunk-8-1.png" width="100%"></p>
</div>
<div class="section level3">
<h3 id="signature-content">3.3 Signature content<a class="anchor" aria-label="anchor" href="#signature-content"></a>
</h3>
<p>Note that there is a difference between the two ways to obtain
molecule names of specific signatures. The function
<code><a href="../reference/getMolecules.html">getGenes()</a></code> primarily refers to the gene names in the count
matrix. It extracts all gene names and then subsets them based on the
input for the <code>signatures</code> argument, if provided. This method
is useful when you need to work with the actual genes present in your
count matrix. The function <code><a href="../reference/getSignature.html">getGeneSet()</a></code> extracts the genes
of a gene set directly. This is particularly important if the gene set
of interest contains genes that are not found in the count matrix
because they were removed during processing or did not exist in the
count matrix in the first place.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># extracts genes for which data exist in the SPATA2 object</span></span>
<span><span class="co"># that are part of the HM_HYPOXIA gene set</span></span>
<span><span class="va">genes_from_mtr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getMolecules.html">getGenes</a></span><span class="op">(</span><span class="va">object_t269</span>, signature <span class="op">=</span> <span class="st">"HM_HYPOXIA"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">genes_from_mtr</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  chr [1:121] "GBE1" "ENO2" "HK2" "SLC2A1" "P4HA1" "ADM" "P4HA2" "PFKFB3" ...</span></span></code></pre>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># extracts all gene names that make up the HM_HYPOXIA gene set</span></span>
<span><span class="co"># regardless of whether data (counts) exist</span></span>
<span><span class="va">genes_directly</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getSignature.html">getGeneSet</a></span><span class="op">(</span><span class="va">object_t269</span>, <span class="st">"HM_HYPOXIA"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">genes_directly</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  chr [1:188] "PGK1" "PDK1" "GBE1" "PFKL" "ALDOA" "ENO2" "PGM1" "NDRG1" ...</span></span></code></pre>
<p>Note, that <code>genes_directly</code> contains more genes, because
it also includes genes that are not present in the data set.</p>
</div>
</div>
<div class="section level2">
<h2 id="multiple-assays">4. Multiple assays<a class="anchor" aria-label="anchor" href="#multiple-assays"></a>
</h2>
<p>As an example we can download a SPATA2 object hat has been created
with a <a href="https://www.10xgenomics.com/datasets/visium-cytassist-gene-and-protein-expression-library-of-human-tonsil-with-add-on-antibodies-h-e-6-5-mm-ffpe-2-standard" class="external-link">dataset</a>
that contains both gene- and protein expression.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SPATAData</span><span class="op">)</span></span>
<span><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/downloadSpataObject.html">downloadSpataObject</a></span><span class="op">(</span><span class="st">"HumanTonsilGP"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"data/object_protein_and_gene.RDS"</span><span class="op">)</span></span></code></pre></div>
<p>The assays are of modality <em>gene</em> and <em>protein</em>.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/methods/show.html" class="external-link">show</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class SPATA2</span></span>
<span><span class="co">## Sample: HumanTonsilGP </span></span>
<span><span class="co">## Size: 4194 x 18119 (spots x molecules)</span></span>
<span><span class="co">## Memory: 562.98 Mb</span></span>
<span><span class="co">## Platform: VisiumSmall </span></span>
<span><span class="co">## Molecular assays (2):</span></span>
<span><span class="co">## 1. Assay</span></span>
<span><span class="co">##  Molecular modality: gene</span></span>
<span><span class="co">##  Distinct molecules: 18085</span></span>
<span><span class="co">##  Matrices (1):</span></span>
<span><span class="co">##  -counts (active)</span></span>
<span><span class="co">## 2. Assay</span></span>
<span><span class="co">##  Molecular modality: protein</span></span>
<span><span class="co">##  Distinct molecules: 34</span></span>
<span><span class="co">##  Matrices (1):</span></span>
<span><span class="co">##  -counts (active)</span></span>
<span><span class="co">## Registered images (2):</span></span>
<span><span class="co">## - hires (2000x1634 px, loaded)</span></span>
<span><span class="co">## - lowres (600x490 px, active, loaded)</span></span>
<span><span class="co">## Meta variables (2): sample, tissue_section</span></span></code></pre>
<p>Note that gene names and protein names can be equal. For SPATA2 to
differentiate between them they must be named differently. When created
with <code><a href="../reference/initiateSpataObjectVisium.html">initiateSpataObjectVisium()</a></code>, protein names are forced
into lower case.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">proteins</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getMolecules.html">getProteins</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">proteins</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  chr [1:34] "cd163" "cr2" "pcna" "vim" "krt5" "cd68" "ceacam8" "ptprc" ...</span></span></code></pre>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getMolecules.html">getGenes</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">genes</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  chr [1:18085] "SAMD11" "NOC2L" "KLHL17" "PLEKHN1" "PERM1" "HES4" "ISG15" ...</span></span></code></pre>
<p>This allows to refer to either gene or protein more conveniently and
unambiguously.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot left - gene</span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object</span>, color_by <span class="op">=</span> <span class="st">"cd4"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot right - protein</span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object</span>, color_by <span class="op">=</span> <span class="st">"CD4"</span><span class="op">)</span></span></code></pre></div>
<p><img src="molecular-variables_files/figure-html/unnamed-chunk-14-1.png" width="50%"><img src="molecular-variables_files/figure-html/unnamed-chunk-14-2.png" width="50%"></p>
<p>If you want to extract data matrices, you can rely on the
<code><a href="../reference/activateAssay.html">activeAssay()</a></code> system. Alternatively, you can always specify
the assay of interest with the argument <code>assay_name</code>.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># currently gene</span></span>
<span><span class="fu"><a href="../reference/activateAssay.html">activeAssay</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "gene"</span></span></code></pre>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># gene counts</span></span>
<span><span class="va">gcount_mtr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getCountMatrix.html">getCountMatrix</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">gcount_mtr</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Formal class 'dgCMatrix' [package "Matrix"] with 6 slots</span></span>
<span><span class="co">##   ..@ i       : int [1:34215510] 11 12 16 17 19 20 21 23 25 26 ...</span></span>
<span><span class="co">##   ..@ p       : int [1:4195] 0 8790 17541 24451 32174 40302 49340 57399 66681 74474 ...</span></span>
<span><span class="co">##   ..@ Dim     : int [1:2] 18085 4194</span></span>
<span><span class="co">##   ..@ Dimnames:List of 2</span></span>
<span><span class="co">##   .. ..$ : chr [1:18085] "SAMD11" "NOC2L" "KLHL17" "PLEKHN1" ...</span></span>
<span><span class="co">##   .. ..$ : chr [1:4194] "AACACGTGCATCGCAC-1" "AACACTTGGCAAGGAA-1" "AACAGGAAGAGCATAG-1" "AACAGGATTCATAGTT-1" ...</span></span>
<span><span class="co">##   ..@ x       : num [1:34215510] 1 1 1 1 1 4 4 3 3 18 ...</span></span>
<span><span class="co">##   ..@ factors : list()</span></span></code></pre>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># protein counts</span></span>
<span><span class="va">pcount_mtr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getCountMatrix.html">getCountMatrix</a></span><span class="op">(</span><span class="va">object</span>, assay_name <span class="op">=</span> <span class="st">"protein"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">pcount_mtr</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Formal class 'dgCMatrix' [package "Matrix"] with 6 slots</span></span>
<span><span class="co">##   ..@ i       : int [1:142550] 0 1 2 3 4 5 6 7 8 9 ...</span></span>
<span><span class="co">##   ..@ p       : int [1:4195] 0 34 68 102 136 170 204 238 272 306 ...</span></span>
<span><span class="co">##   ..@ Dim     : int [1:2] 34 4194</span></span>
<span><span class="co">##   ..@ Dimnames:List of 2</span></span>
<span><span class="co">##   .. ..$ : chr [1:34] "cd163" "cr2" "pcna" "vim" ...</span></span>
<span><span class="co">##   .. ..$ : chr [1:4194] "AACACGTGCATCGCAC-1" "AACACTTGGCAAGGAA-1" "AACAGGAAGAGCATAG-1" "AACAGGATTCATAGTT-1" ...</span></span>
<span><span class="co">##   ..@ x       : num [1:142550] 19675 35856 57685 48611 4606 ...</span></span>
<span><span class="co">##   ..@ factors : list()</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="conclusion">5. Conclusion<a class="anchor" aria-label="anchor" href="#conclusion"></a>
</h2>
<p>For all functions presented in this tutorial, there exists a general
function as well as wrappers for specific modalities. The general
function always works but requires specifying the assay of interest with
the argument <code>assay_name</code>.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Jan Kueckelhaus, Dieter-Henrik Heiland.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
