<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Spatial Segmentation • SPATA2</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Spatial Segmentation">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SPATA2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">3.1.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li>
  <a href="../articles/spata-data.html">SPATAData</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Initiating a SPATA2 object</li>
    <li>
      <a href="../articles/initiation-and-preprocessing-customized.html">Initiation &amp; Preprocessing with individual data</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-merfish.html">Initiation &amp; Preprocessing with MERFISH data</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-visium.html">Initiation &amp; Preprocessing with Visium data</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-visium-hd.html">Initiation &amp; Preprocessing with VisiumHD data</a>
    </li>
    <li class="dropdown-header">Basic Manipulation of a SPATA2 object</li>
    <li>
      <a href="../articles/adding-data.html">Adding Data to SPATA2 objects</a>
    </li>
    <li>
      <a href="../articles/package-compatibility.html">Platform Compatibility</a>
    </li>
    <li>
      <a href="../articles/molecular-variables.html">Molecular Variables</a>
    </li>
    <li>
      <a href="../articles/image-handling.html">Image Handling</a>
    </li>
    <li>
      <a href="../articles/spatial-segmentation.html">Spatial Segmentation</a>
    </li>
    <li class="dropdown-header">Implemented Algorithms &amp; Concepts</li>
    <li>
      <a href="../articles/cnv.html">Copy Number Variations</a>
    </li>
    <li>
      <a href="../articles/clustering.html">Clustering</a>
    </li>
    <li>
      <a href="../articles/dimensional-reduction.html">Dimensional Reduction</a>
    </li>
    <li>
      <a href="../articles/dea.html">Differential Expression</a>
    </li>
    <li>
      <a href="../articles/gsea.html">Gene Set Enrichment</a>
    </li>
    <li>
      <a href="../articles/spatial-measures.html">Spatial Measures</a>
    </li>
    <li class="dropdown-header">Spatial Gradient Screening</li>
    <li>
      <a href="../articles/creating-spatial-annotations.html">Creating Spatial Annotations</a>
    </li>
    <li>
      <a href="../articles/using-spatial-annotations.html">Using Spatial Annotations</a>
    </li>
    <li>
      <a href="../articles/spatial-annotation-screening.html">Spatial Annotation Screening</a>
    </li>
    <li>
      <a href="../articles/creating-spatial-trajectories.html">Creating Spatial Trajectories</a>
    </li>
    <li>
      <a href="../articles/spatial-trajectory-screening.html">Spatial Trajectory Screening</a>
    </li>
    <li class="dropdown-header">Visualization</li>
    <li>
      <a href="../articles/plotting-surface.html">Surface Plotting</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/newslog.html">Newslog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Spatial Segmentation</h1>
            
      

      <div class="hidden name"><code>spatial-segmentation.Rmd</code></div>

    </div>

    
    
<div class="section level3">
<h3 id="introduction">1. Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h3>
<p>Spatial transcriptomic samples come along with an underlying
histology image. To integrate the histological classification of the
area a specific observations covers, the sample can be manually
segmented with SPATA2. As an example we are using a spatial
transcriptomic sample of a central nervous system malignancy that
features three different, adjacent histological areas: Tumor, a
transition zone as well as infiltrated cortex.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load required packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://themilolab.com/" class="external-link">SPATA2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># load SPATA2 inbuilt data </span></span>
<span><span class="va">object_t269</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadExampleObject.html">loadExampleObject</a></span><span class="op">(</span><span class="st">"UKF269T"</span>, meta <span class="op">=</span> <span class="cn">TRUE</span>, process <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="create-spatial-segmentation-variables">2. Create spatial segmentation variables<a class="anchor" aria-label="anchor" href="#create-spatial-segmentation-variables"></a>
</h2>
<p>Spatial segmentation, or sometimes referred to as manual annotation,
means to create grouping variables that assign a label to each
observation based on the histomorphological features of the image. In
the example object of sample UKF269T, we provide such a variable. We
just named it <em>histology</em>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># variables that have been created using spatial segmentation</span></span>
<span><span class="co"># curently only one: "histology"</span></span>
<span><span class="fu"><a href="../reference/getSpatSegmVarNames.html">getSpatSegmVarNames</a></span><span class="op">(</span><span class="va">object_t269</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "histology"</span></span></code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot the histology image in coordinates frame</span></span>
<span><span class="fu"><a href="../reference/plotImage.html">plotImage</a></span><span class="op">(</span><span class="va">object_t269</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="../reference/ggpLayerFrameByCoords.html">ggpLayerFrameByCoords</a></span><span class="op">(</span><span class="va">object_t269</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># grouping variable created with spatial segmentation</span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object_t269</span>, color_by <span class="op">=</span> <span class="st">"histology"</span><span class="op">)</span> </span></code></pre></div>
<p><img src="spatial-segmentation_files/figure-html/unnamed-chunk-2-1.png" width="50%"><img src="spatial-segmentation_files/figure-html/unnamed-chunk-2-2.png" width="50%"></p>
<p>To create a spatial segmentation variabl as displayed in the figure
above yourself, use the function
<code><a href="../reference/createSpatialSegmentation.html">createSpatialSegmentation()</a></code>. This animation exemplifies the
workflow:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">object_t269</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createSpatialSegmentation.html">createSpatialSegmentation</a></span><span class="op">(</span><span class="va">object_t269</span><span class="op">)</span></span></code></pre></div>
<video width="640" height="480" controls><source src="../reference/figures/create_spatial_segmentation.mp4" type="video/mp4" srcset="NA"></source></video><p>
</p>
<p>To create a new segmentation variable click on ‘Create new
segmentation variable’. You are then prompted to enter the name that you
want to give the new variable. Here, we simply call it
<em>example_spat_segm</em>. After clicking on ‘Add segmentation variable
a new’ variable is created in the meta data.frame and all of it’s values
are ‘unnamed’. This is the default, has we haven’t segmented anything
yet. Working on segmentation variables means to consecutively label all
observations depending on the histological area they cover. To achieve
this, encircle the areas that you want to label on the Interaction-plot.
Start drawing by double clicking on the plot or with the shortcut on
your keyboard d. If you are drawing the plot tells you. By
double-clicking again you leave the drawing mode. This can be used to
zoom in or out to reorientate. If you double-click again the endpoint of
the line drawn will connect to the position of your cursor and you can
continue drawing. Encircle the area till start and endpoint of the line
are close enough to be connected such that the circle closes. To close
the circle click on ‘Highlight’. This highlights the observations that
are encircled this way. After highlighting the observations the name
that you want to label them with can be entered below the
‘Highlight’-button. Click on ‘Name’ to save the results. The results
should be immediately displayed in the plot on the left. This can be
done over and over again till you are satisfied. Note that the label you
gave a observation is overwritten if you include it again in another
area.</p>
<p>If you run the function with your own <code>SPATA2</code> object for
the first time, there is no plot displayed on the left. This is because
there is currently no segmentation variable in the meta data.frame that
can be worked on. In this example, we have already created the spatial
segmentation variable <em>histology</em>. And do not bother to much with
single data points your drawn polygons omit. The function
<code><a href="../reference/dissolveGroups.html">dissolveGroups()</a></code> might spare you time and trouble.</p>
<p>Furthermore, as the video below exemplifies, you can draw multiple
outlines at the same time and give them one single label. On top of
that, you can draw “holes” inside outlines. Note that outlines must not
intersect!</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">object_t313</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadExampleObject.html">loadExampleObject</a></span><span class="op">(</span><span class="st">"UKFT313"</span>, process <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">object_t313</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createSpatialSegmentation.html">createSpatialSegmentation</a></span><span class="op">(</span><span class="va">object_t313</span><span class="op">)</span></span></code></pre></div>
<video width="640" height="480" controls><source src="../reference/figures/create_spatial_segmentation_advanced.mp4" type="video/mp4" srcset="NA"></source></video><p>
</p>
</div>
<div class="section level2">
<h2 id="data-storage">3. Data storage<a class="anchor" aria-label="anchor" href="#data-storage"></a>
</h2>
<p>Grouping variables in transcriptomic studies are often created by
cluster algorithms and are stored in form of factors in the meta
data.frame. As spatial segmentation variables are grouping variables,
too, they are stored there as well.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># show all grouping (factor) variables in the meta data.frame</span></span>
<span><span class="fu"><a href="../reference/getMetaDf.html">getMetaDf</a></span><span class="op">(</span><span class="va">object_t269</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">barcodes</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/where.html" class="external-link">where</a></span><span class="op">(</span><span class="va">is.factor</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 3,213 × 6</span></span></span>
<span><span class="co">##    barcodes           tissue_section   seurat_clusters histology  bayes_space</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>      </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> GTAGCGCTGTTGTAGT-1 tissue_section_1 1               tumor      2          </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> TTGTTTGTGTAAATTC-1 tissue_section_1 1               tumor      2          </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> CGTAGCGCCGACGTTG-1 tissue_section_1 2               tumor      2          </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> GTAGACAACCGATGAA-1 tissue_section_1 2               tumor      2          </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> ACAGATTAGGTTAGTG-1 tissue_section_1 9               tumor      2          </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> TGAGATCAAATACTCA-1 tissue_section_1 2               tumor      4          </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> CTGGTCCTAACTTGGC-1 tissue_section_1 9               tumor      4          </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> TGCACGAGTCGGCAGC-1 tissue_section_1 7               tumor      4          </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> ATAGTCTTTGACGTGC-1 tissue_section_1 9               transition 4          </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> GGGTGGTCCAGCCTGT-1 tissue_section_1 4               transition 4          </span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 3,203 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 1 more variable: example_spat_segm &lt;fct&gt;</span></span></span></code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># obtain only names from variables created with spatial segmentation</span></span>
<span><span class="va">spat_segm_vars</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getSpatSegmVarNames.html">getSpatSegmVarNames</a></span><span class="op">(</span><span class="va">object_t269</span><span class="op">)</span></span>
<span></span>
<span><span class="va">spat_segm_vars</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "histology"         "example_spat_segm"</span></span></code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># show only variables that have been created by spatial segmentation</span></span>
<span><span class="fu"><a href="../reference/getMetaDf.html">getMetaDf</a></span><span class="op">(</span><span class="va">object_t269</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">barcodes</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/all_of.html" class="external-link">all_of</a></span><span class="op">(</span><span class="va">spat_segm_vars</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 3,213 × 3</span></span></span>
<span><span class="co">##    barcodes           histology  example_spat_segm</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>            </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> GTAGCGCTGTTGTAGT-1 tumor      unnamed          </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> TTGTTTGTGTAAATTC-1 tumor      unnamed          </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> CGTAGCGCCGACGTTG-1 tumor      unnamed          </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> GTAGACAACCGATGAA-1 tumor      unnamed          </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> ACAGATTAGGTTAGTG-1 tumor      unnamed          </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> TGAGATCAAATACTCA-1 tumor      unnamed          </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> CTGGTCCTAACTTGGC-1 tumor      unnamed          </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> TGCACGAGTCGGCAGC-1 tumor      unnamed          </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> ATAGTCTTTGACGTGC-1 transition unnamed          </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> GGGTGGTCCAGCCTGT-1 transition unnamed          </span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 3,203 more rows</span></span></span></code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># visualize the (unfinished) spatial segmentation variable </span></span>
<span><span class="co"># from the gif above</span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object_t269</span>, color_by <span class="op">=</span> <span class="st">"example_spat_segm"</span>, clrp_adjust <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"unnamed"</span> <span class="op">=</span> <span class="st">"lightgrey"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="spatial-segmentation_files/figure-html/unnamed-chunk-7-1.png" width="50%"></p>
</div>
<div class="section level2">
<h2 id="working-with-segmentation-variables">4. Working with segmentation variables<a class="anchor" aria-label="anchor" href="#working-with-segmentation-variables"></a>
</h2>
<p>Segmentation variables can be used like any other grouping variable
and are usually referred to via the arguments <code>group_by</code>,
<code>grouping_variable</code> or <code>across</code>.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># histology variable is a grouping option ...</span></span>
<span><span class="fu"><a href="../reference/getGroupingOptions.html">getGroupingOptions</a></span><span class="op">(</span><span class="va">object_t269</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##              factor              factor              factor              factor </span></span>
<span><span class="co">##    "tissue_section"   "seurat_clusters"         "histology"       "bayes_space" </span></span>
<span><span class="co">##              factor </span></span>
<span><span class="co">## "example_spat_segm"</span></span></code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># but not all grouping options derive from spatial segmentation</span></span>
<span><span class="fu"><a href="../reference/getSpatSegmVarNames.html">getSpatSegmVarNames</a></span><span class="op">(</span><span class="va">object_t269</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "histology"         "example_spat_segm"</span></span></code></pre>
<p>Since the new spatial segmentation variable is a meta feature as any
other grouping variables the same rules and options apply to it
<code><a href="../reference/getGroupNames.html">getGroupNames()</a></code>, <code><a href="../reference/renameGroups.html">renameGroups()</a></code>,
<code><a href="../reference/relevelGroups.html">relevelGroups()</a></code>.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># check group (segment) names</span></span>
<span><span class="fu"><a href="../reference/getGroupNames.html">getGroupNames</a></span><span class="op">(</span><span class="va">object_t269</span>, grouping <span class="op">=</span> <span class="st">"histology"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "tumor"       "transition"  "infiltrated"</span></span></code></pre>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># rename groups (only temporary without storing the results)</span></span>
<span><span class="fu"><a href="../reference/renameGroups.html">renameGroups</a></span><span class="op">(</span><span class="va">object_t269</span>, grouping_variable <span class="op">=</span> <span class="st">"histology"</span>, <span class="st">"tumor_renamed"</span> <span class="op">=</span> <span class="st">"tumor"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">.</span>, color_by <span class="op">=</span> <span class="st">"histology"</span>, pt_clrp <span class="op">=</span> <span class="st">"npg"</span><span class="op">)</span></span></code></pre></div>
<p><img src="spatial-segmentation_files/figure-html/unnamed-chunk-9-1.png" width="50%"></p>
<p>Spatial segmentation variables can be also used to conduct <a href="dea.html">DEA</a> and <a href="gsea.html">GSEA</a> based on
histology and access results in plotting functions.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># run DEA based on histology</span></span>
<span><span class="va">object_t269</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runDEA.html">runDEA</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">object_t269</span>, across <span class="op">=</span> <span class="st">"histology"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotDeaVolcano.html">plotDeaVolcano</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">object_t269</span>,</span>
<span>  across <span class="op">=</span> <span class="st">"histology"</span>, </span>
<span>  use_pseudolog <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>  label_genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"EGFR"</span>, <span class="st">"MBP"</span>, <span class="st">"SNAP25"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="spatial-segmentation_files/figure-html/unnamed-chunk-11-1.png" width="100%"></p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSurfaceComparison.html">plotSurfaceComparison</a></span><span class="op">(</span><span class="va">object_t269</span>, color_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"EGFR"</span>, <span class="st">"MBP"</span>, <span class="st">"SNAP25"</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="spatial-segmentation_files/figure-html/unnamed-chunk-12-1.png" width="100%"></p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotBoxplot.html">plotBoxplot</a></span><span class="op">(</span><span class="va">object_t269</span>, variables <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"EGFR"</span>, <span class="st">"MBP"</span>, <span class="st">"SNAP25"</span><span class="op">)</span>, across <span class="op">=</span> <span class="st">"histology"</span>, nrow <span class="op">=</span> <span class="fl">1</span>, clrp <span class="op">=</span> <span class="st">"npg"</span><span class="op">)</span></span></code></pre></div>
<p><img src="spatial-segmentation_files/figure-html/unnamed-chunk-13-1.png" width="100%"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Jan Kueckelhaus, Dieter-Henrik Heiland.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
