<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>SPATAData â€¢ SPATA2</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="SPATAData">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SPATA2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">3.1.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li>
  <a href="../articles/spata-data.html">SPATAData</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Initiating a SPATA2 object</li>
    <li>
      <a href="../articles/initiation-and-preprocessing-customized.html">Initiation &amp; Preprocessing with individual data</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-merfish.html">Initiation &amp; Preprocessing with MERFISH data</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-visium.html">Initiation &amp; Preprocessing with Visium data</a>
    </li>
    <li>
      <a href="../articles/initiation-and-preprocessing-visium-hd.html">Initiation &amp; Preprocessing with VisiumHD data</a>
    </li>
    <li class="dropdown-header">Basic Manipulation of a SPATA2 object</li>
    <li>
      <a href="../articles/adding-data.html">Adding Data to SPATA2 objects</a>
    </li>
    <li>
      <a href="../articles/package-compatibility.html">Platform Compatibility</a>
    </li>
    <li>
      <a href="../articles/molecular-variables.html">Molecular Variables</a>
    </li>
    <li>
      <a href="../articles/image-handling.html">Image Handling</a>
    </li>
    <li>
      <a href="../articles/spatial-segmentation.html">Spatial Segmentation</a>
    </li>
    <li class="dropdown-header">Implemented Algorithms &amp; Concepts</li>
    <li>
      <a href="../articles/cnv.html">Copy Number Variations</a>
    </li>
    <li>
      <a href="../articles/clustering.html">Clustering</a>
    </li>
    <li>
      <a href="../articles/dimensional-reduction.html">Dimensional Reduction</a>
    </li>
    <li>
      <a href="../articles/dea.html">Differential Expression</a>
    </li>
    <li>
      <a href="../articles/gsea.html">Gene Set Enrichment</a>
    </li>
    <li>
      <a href="../articles/spatial-measures.html">Spatial Measures</a>
    </li>
    <li class="dropdown-header">Spatial Gradient Screening</li>
    <li>
      <a href="../articles/creating-spatial-annotations.html">Creating Spatial Annotations</a>
    </li>
    <li>
      <a href="../articles/using-spatial-annotations.html">Using Spatial Annotations</a>
    </li>
    <li>
      <a href="../articles/spatial-annotation-screening.html">Spatial Annotation Screening</a>
    </li>
    <li>
      <a href="../articles/creating-spatial-trajectories.html">Creating Spatial Trajectories</a>
    </li>
    <li>
      <a href="../articles/spatial-trajectory-screening.html">Spatial Trajectory Screening</a>
    </li>
    <li class="dropdown-header">Visualization</li>
    <li>
      <a href="../articles/plotting-surface.html">Surface Plotting</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/newslog.html">Newslog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>SPATAData</h1>
            
      

      <div class="hidden name"><code>spata-data.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">1. Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The package SPATAData gives access to our data base of spatial
transcriptomic samples. Furthermore, it provides easy access to data
sets that have been already published. We are continuously updating this
collection so make sure to check for package updates on a regular
basis.</p>
<p>Please note that many of these data sets are not owned by us! Make
sure to use the correct citation if you download and use them for your
analysis. See more under section 4. Citation.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install SPATAData with:</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span>repo <span class="op">=</span> <span class="st">"theMILOlab/SPATAData"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># and if you have not already:  </span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"kueckelj/confuns"</span><span class="op">)</span></span></code></pre></div>
<p>We, the <strong>MILOlab</strong>, are a workgroup focused on
neurooncology. Our database is predominantly composed of human samples
from the human cerebrum (panel a). Nonetheless, we have also curated
multiple objects from various other organs with distinct histological
classifications (panel b). Utilize the source <code>data.frame</code> as
described below to obtain an overview and filter samples that may be
relevant to your research. Note the differentiation between the organ
<em>Brain</em> for mice tissue donors and <em>Cerebrum</em> for human
tissue donors. This distinction is important because, in the case of
Visium datasets, mouse samples usually encompass the entire intracranial
central nervous system (commonly referred to as <em>Brain</em>). Due to
their size, human brain samples are derived from specific organs
(column: <em>organ</em>, with values such as <em>Cerebrum</em>,
<em>Midbrain</em>, <em>Cerebellum</em>) and specific locations (column:
<em>organ_part</em>, with values such as <em>frontal lobe</em>,
<em>temporal lobe</em>, <em>corpus callosum</em>). While the number of
platforms (panel c) is currently dominated by <code>VisiumSmall</code>
we aim to expand this collection of data sets in the future to include
more of everything.</p>
<p><img src="spata-data_files/figure-html/unnamed-chunk-2-1.png" width="100%"></p>
</div>
<div class="section level2">
<h2 id="the-source-data-frame">2. The source data.frame<a class="anchor" aria-label="anchor" href="#the-source-data-frame"></a>
</h2>
<p>The last version of SPATAData used to have an interactive interface
in which data samples could be viewed and downloaded by mouse click.
This interface is currently not available (but will be, hopefully, in
the months to come.) Till then, you can make use of the source
data.frame directly in combination with some <code>dplyr</code> logic.
The source data.frame of SPATAData, as obtained by
<code><a href="https://rdrr.io/pkg/SPATAData/man/sourceDataFrame.html" class="external-link">sourceDataFrame()</a></code>, contains web links as well as meta data
to multiple spatial data sets that have been published so far. Currently
it counts a total of 195 samples across 20 organs and 19 histological
classifications. In the source data.frame every row corresponds to a
data set as stored in a <code>SPATA2</code> object. Hence, you can use
dplyr to filter for data sets that fit your interest by filtering for
specific characteristics. Please refer to the documentationo of
<code><a href="https://rdrr.io/pkg/SPATAData/man/sourceDataFrame.html" class="external-link">sourceDataFrame()</a></code> to familiarize yourself with the
variables of the source data.frame and the kind of meta data they
carry.</p>
<p>You can use <code><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique()</a></code> on each non numeric variable to
obtain groups by which to filter the object.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load required packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://themilolab.com/" class="external-link">SPATA2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SPATAData</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://stringr.tidyverse.org" class="external-link">stringr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">#assign the data.frame</span></span>
<span><span class="va">source_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SPATAData/man/sourceDataFrame.html" class="external-link">sourceDataFrame</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># get unique donor species types</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">source_df</span><span class="op">$</span><span class="va">donor_species</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Homo sapiens" "Mus musculus"</span></span></code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get the different organs for which data exists</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">source_df</span><span class="op">$</span><span class="va">organ</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "Breast"        "Cerebrum"      "Tonsil"        "Lung"         </span></span>
<span><span class="co">##  [5] "Pancreas"      "Colon"         "Kidney"        "Ovary"        </span></span>
<span><span class="co">##  [9] "Prostate"      "Brain"         "Heart"         "Thyroid Gland"</span></span>
<span><span class="co">## [13] "Brainstem"     "Cerebellum"    "Lymph Node"    "Spinal Cord"  </span></span>
<span><span class="co">## [17] "Rectum"        "Liver"         "Skin"          "Uterus"</span></span></code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get additional specifications of anatomical location</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">source_df</span><span class="op">$</span><span class="va">organ_part</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] NA                "frontal"         "parietal"        "frontotemporal" </span></span>
<span><span class="co">##  [5] "occipital"       "temporal"        "prefrontal"      "hemisphere"     </span></span>
<span><span class="co">##  [9] "thalamus"        "corpus_callosum" "Ascending"       "Cecum"          </span></span>
<span><span class="co">## [13] "Sigmoid"         "Unspecified"</span></span></code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get the different histo subclasses for which data exists</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">source_df</span><span class="op">$</span><span class="va">histo_class</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] NA                                          </span></span>
<span><span class="co">##  [2] "Glioblastoma"                              </span></span>
<span><span class="co">##  [3] "Oligodendroglioma"                         </span></span>
<span><span class="co">##  [4] "Astrocytoma"                               </span></span>
<span><span class="co">##  [5] "Cardiac Muscle"                            </span></span>
<span><span class="co">##  [6] "Healthy tissue"                            </span></span>
<span><span class="co">##  [7] "Graves disease"                            </span></span>
<span><span class="co">##  [8] "Hashimoto's thyroiditis and thyroid nodule"</span></span>
<span><span class="co">##  [9] "Hashimoto's thyroiditis"                   </span></span>
<span><span class="co">## [10] "Cortex"                                    </span></span>
<span><span class="co">## [11] "Giant cell glioblastoma"                   </span></span>
<span><span class="co">## [12] "Gliosarcoma"                               </span></span>
<span><span class="co">## [13] "Astrocytoma IDH mutated"                   </span></span>
<span><span class="co">## [14] "Ganglioglioma"                             </span></span>
<span><span class="co">## [15] "Diffuse midline glioma, H3K27M-mutant"     </span></span>
<span><span class="co">## [16] "Lymph Node"                                </span></span>
<span><span class="co">## [17] "Spinal Cord"                               </span></span>
<span><span class="co">## [18] "Kidney"                                    </span></span>
<span><span class="co">## [19] "Adenocarcinoma"</span></span></code></pre>
<p>To filter the source data.frame it takes logical tests that represent
your idea of the dataset you need. For instance, if you want all
glioblastoma samples from the frontal and temporal lobe the code would
look like this:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># filter for frontal and temporal glioblastoma</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SPATAData/man/sourceDataFrame.html" class="external-link">sourceDataFrame</a></span><span class="op">(</span><span class="va">histo_class</span> <span class="op">==</span> <span class="st">"Glioblastoma"</span> <span class="op">&amp;</span> <span class="va">organ_part</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"frontal"</span>, <span class="st">"temporal"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 24 Ã— 32</span></span></span>
<span><span class="co">##    sample_name comment donor_id donor_species grade histo_class  institution    </span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> MGH258      <span style="color: #BB0000;">NA</span>      MGH258   Homo sapiens  IV    Glioblastoma Massachusetts â€¦</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> ZH1019_inf  <span style="color: #BB0000;">NA</span>      ZH1007   Homo sapiens  IV    Glioblastoma University Hosâ€¦</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> ZH1019_T1   <span style="color: #BB0000;">NA</span>      ZH1007   Homo sapiens  IV    Glioblastoma University Hosâ€¦</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> ZH881_inf   <span style="color: #BB0000;">NA</span>      ZH881    Homo sapiens  IV    Glioblastoma University Hosâ€¦</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> ZH881_T1    <span style="color: #BB0000;">NA</span>      ZH881    Homo sapiens  IV    Glioblastoma University Hosâ€¦</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> ZH916_bulk  <span style="color: #BB0000;">NA</span>      ZH916    Homo sapiens  IV    Glioblastoma University Hosâ€¦</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> ZH916_inf   <span style="color: #BB0000;">NA</span>      ZH916    Homo sapiens  IV    Glioblastoma University Hosâ€¦</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> ZH916_T1    <span style="color: #BB0000;">NA</span>      ZH916    Homo sapiens  IV    Glioblastoma University Hosâ€¦</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> UKF242T     <span style="color: #BB0000;">NA</span>      UKF242   Homo sapiens  IV    Glioblastoma University Cliâ€¦</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> UKF243T     <span style="color: #BB0000;">NA</span>      UKF243   Homo sapiens  IV    Glioblastoma University Cliâ€¦</span></span>
<span><span class="co">## <span style="color: #949494;"># â„¹ 14 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># â„¹ 25 more variables: lm_source &lt;dttm&gt;, organ &lt;chr&gt;, organ_part &lt;chr&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   organ_side &lt;chr&gt;, pathology &lt;chr&gt;, platform &lt;chr&gt;, pub_citation &lt;chr&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   pub_doi &lt;chr&gt;, pub_journal &lt;chr&gt;, pub_year &lt;dbl&gt;, sex &lt;chr&gt;, source &lt;chr&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   tags &lt;chr&gt;, tissue_age &lt;dbl&gt;, web_link &lt;chr&gt;, workgroup &lt;chr&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   mean_counts &lt;dbl&gt;, median_counts &lt;dbl&gt;, modality_gene &lt;lgl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   modality_metabolite &lt;lgl&gt;, modality_protein &lt;lgl&gt;, n_obs &lt;int&gt;, â€¦</span></span></span></code></pre>
<p>If you want samples from a specific publication:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># look for publications and journals with string subsetting</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SPATAData/man/sourceDataFrame.html" class="external-link">sourceDataFrame</a></span><span class="op">(</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_detect.html" class="external-link">str_detect</a></span><span class="op">(</span><span class="va">pub_citation</span>, pattern <span class="op">=</span> <span class="st">"^Kuppe"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 27 Ã— 27</span></span></span>
<span><span class="co">##    sample_name donor_id donor_species histo_class    histo_class_sub institution</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> 10X001      P1       Homo sapiens  Cardiac Muscle fibrotic_zone   Universityâ€¦</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> 10X0017     P3       Homo sapiens  Cardiac Muscle ischaemic_zone  Universityâ€¦</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> 10X0018     P2       Homo sapiens  Cardiac Muscle border_zone     Universityâ€¦</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> 10X0020     P3       Homo sapiens  Cardiac Muscle remote_zone     Universityâ€¦</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> 10X0025     P2       Homo sapiens  Cardiac Muscle ischaemic_zone  Universityâ€¦</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> 10X0026     P3       Homo sapiens  Cardiac Muscle border_zone     Universityâ€¦</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> 10X0027     P5       Homo sapiens  Cardiac Muscle fibrotic_zone   Universityâ€¦</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> 10X009      P4       Homo sapiens  Cardiac Muscle control         Universityâ€¦</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> ACH0010     P13      Homo sapiens  Cardiac Muscle control         Universityâ€¦</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> ACH0011     P15      Homo sapiens  Cardiac Muscle border_zone     Universityâ€¦</span></span>
<span><span class="co">## <span style="color: #949494;"># â„¹ 17 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># â„¹ 21 more variables: lm_source &lt;dttm&gt;, organ &lt;chr&gt;, pathology &lt;chr&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   platform &lt;chr&gt;, pub_citation &lt;chr&gt;, pub_doi &lt;chr&gt;, pub_journal &lt;chr&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   pub_year &lt;dbl&gt;, source &lt;chr&gt;, tags &lt;chr&gt;, web_link &lt;chr&gt;, workgroup &lt;chr&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   mean_counts &lt;dbl&gt;, median_counts &lt;dbl&gt;, modality_gene &lt;lgl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   modality_metabolite &lt;lgl&gt;, modality_protein &lt;lgl&gt;, n_obs &lt;int&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   n_tissue_sections &lt;int&gt;, obs_unit &lt;chr&gt;, obj_size &lt;lbstr_by&gt;</span></span></span></code></pre>
<p>The number of conditions is unlimited. Make yourself familiar with
the variables of the source data.frame as well as with the
<code><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">dplyr::filter()</a></code> logic. You can even process the data.frame
to filter for specific queries. E.g. if you want patient wise matching
of samples.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># look for several samples from one single patient</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SPATAData/man/sourceDataFrame.html" class="external-link">sourceDataFrame</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">donor_id</span><span class="op">)</span> <span class="op">&amp;</span> <span class="va">organ</span> <span class="op">==</span> <span class="st">"Cerebrum"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">donor_id</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="co"># count the Cerebrum samples by donor</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>ns_by_donor <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">ns_by_donor</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="co"># keep only those samples with n &gt; 1</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">donor_id</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 30 Ã— 32</span></span></span>
<span><span class="co">## <span style="color: #949494;"># Groups:   donor_id [12]</span></span></span>
<span><span class="co">##    sample_name donor_id donor_species grade histo_class  institution            </span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> UKF242C     UKF242   Homo sapiens  <span style="color: #BB0000;">NA</span>    Cortex       University Clinic Freiâ€¦</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> UKF242T     UKF242   Homo sapiens  IV    Glioblastoma University Clinic Freiâ€¦</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> UKF248C     UKF248   Homo sapiens  <span style="color: #BB0000;">NA</span>    Cortex       University Clinic Freiâ€¦</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> UKF248T     UKF248   Homo sapiens  IV    Glioblastoma University Clinic Freiâ€¦</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> UKF256C     UKF256   Homo sapiens  <span style="color: #BB0000;">NA</span>    Cortex       University Clinic Freiâ€¦</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> UKF256TC    UKF256   Homo sapiens  IV    Cortex       University Clinic Freiâ€¦</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> UKF256TI    UKF256   Homo sapiens  IV    Glioblastoma University Clinic Freiâ€¦</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> UKF259C     UKF259   Homo sapiens  <span style="color: #BB0000;">NA</span>    Cortex       University Clinic Freiâ€¦</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> UKF259T     UKF259   Homo sapiens  IV    Glioblastoma University Clinic Freiâ€¦</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> UKF260T     UKF260   Homo sapiens  IV    Glioblastoma University Clinic Freiâ€¦</span></span>
<span><span class="co">## <span style="color: #949494;"># â„¹ 20 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># â„¹ 26 more variables: lm_source &lt;dttm&gt;, organ &lt;chr&gt;, organ_part &lt;chr&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   organ_side &lt;chr&gt;, pathology &lt;chr&gt;, platform &lt;chr&gt;, pub_citation &lt;chr&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   pub_doi &lt;chr&gt;, pub_journal &lt;chr&gt;, pub_year &lt;dbl&gt;, sex &lt;chr&gt;, source &lt;chr&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   tags &lt;chr&gt;, tissue_age &lt;dbl&gt;, web_link &lt;chr&gt;, workgroup &lt;chr&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   mean_counts &lt;dbl&gt;, median_counts &lt;dbl&gt;, modality_gene &lt;lgl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   modality_metabolite &lt;lgl&gt;, modality_protein &lt;lgl&gt;, n_obs &lt;int&gt;, â€¦</span></span></span></code></pre>
</div>
<div class="section level2">
<h2 id="downloads">3. Downloads<a class="anchor" aria-label="anchor" href="#downloads"></a>
</h2>
<p>Whether you get them by filtering the source data.frame or because
you know them by name, to download <code>SPATA2</code> objects the
sample names are required. There are two functions with which to
download <code>SPATA2</code> object. The downloaded objects are
completely unprocessed. Hence, the plots you see above derive from raw
counts. Refer to the vignettes on object creation and processing to find
the pipeline you see fit for your data samples.</p>
<ul>
<li>
<code><a href="../reference/downloadSpataObject.html">downloadSpataObject()</a></code>: To download single objects. They
can be saved on disc automatically but this function is particularly
equipped for quick downloads and assignment to a variable in your R
session.</li>
<li>
<code><a href="../reference/downloadSpataObjects.html">downloadSpataObjects()</a></code>: Takes a character vector of
sample names, then downloads and stores them all together in the
specified directory.</li>
</ul>
</div>
<div class="section level2">
<h2 id="download-and-assign">3.1 Download and assign<a class="anchor" aria-label="anchor" href="#download-and-assign"></a>
</h2>
<p>This code chunk downloads single objects by sample name. It assigns
the result to a variable in your global environment and you can
immediately start with analysis and visualization.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># download objects by sample name and assign them to environment variables</span></span>
<span><span class="va">object_heart</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/downloadSpataObject.html">downloadSpataObject</a></span><span class="op">(</span>sample_name <span class="op">=</span> <span class="st">"ACH0010"</span><span class="op">)</span></span>
<span><span class="va">object_gbm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/downloadSpataObject.html">downloadSpataObject</a></span><span class="op">(</span>sample_name <span class="op">=</span> <span class="st">"UKF242T"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># left plot</span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object_heart</span>, color_by <span class="op">=</span> <span class="st">"HM_HYPOXIA"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># right plot</span></span>
<span><span class="fu"><a href="../reference/plotSurface.html">plotSurface</a></span><span class="op">(</span><span class="va">object_gbm</span>, color_by <span class="op">=</span> <span class="st">"GFAP"</span>, alpha_by <span class="op">=</span> <span class="st">"GFAP"</span><span class="op">)</span></span></code></pre></div>
<p><img src="spata-data_files/figure-html/unnamed-chunk-8-1.png" width="50%"><img src="spata-data_files/figure-html/unnamed-chunk-8-2.png" width="50%"></p>
</div>
<div class="section level2">
<h2 id="download-and-saving-on-disk">3.2 Download and saving on disk<a class="anchor" aria-label="anchor" href="#download-and-saving-on-disk"></a>
</h2>
<p>This code chunk uses filtering and
<code><a href="../reference/downloadSpataObjects.html">downloadSpataObjects()</a></code> to download a complete set into a
single folder.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># filter source data.frame</span></span>
<span><span class="va">healthy_human_cortex_samples</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/SPATAData/man/sourceDataFrame.html" class="external-link">sourceDataFrame</a></span><span class="op">(</span><span class="va">organ</span> <span class="op">==</span> <span class="st">"Cerebrum"</span> <span class="op">&amp;</span> <span class="va">histo_class</span> <span class="op">==</span> <span class="st">"Cortex"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">sample_name</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># create this directory or adjust it to your liking</span></span>
<span><span class="va">folder</span> <span class="op">&lt;-</span> <span class="st">"spata_objects/healthy_cortex"</span> </span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">folder</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># download samples</span></span>
<span><span class="fu"><a href="../reference/downloadSpataObjects.html">downloadSpataObjects</a></span><span class="op">(</span></span>
<span>  sample_names <span class="op">=</span> <span class="va">healthy_human_cortex_samples</span>, </span>
<span>  folder <span class="op">=</span> <span class="va">folder</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># show results</span></span>
<span><span class="va">healthy_human_cortex_samples</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "151507"   "151508"   "151509"   "151510"   "151669"   "151670"  </span></span>
<span><span class="co">##  [7] "151671"   "151672"   "151673"   "151674"   "151675"   "151676"  </span></span>
<span><span class="co">## [13] "UKF241C"  "UKF242C"  "UKF248C"  "UKF256C"  "UKF256TC" "UKF259C" </span></span>
<span><span class="co">## [19] "UKF265C"  "UKF313C"  "UKF334C"  "GBM5_2"</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="citation">4. Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h2>
<p>As mentioned in the introduction, many samples provided here are not
owned by us, the MILOlab. We have collected these samples from openly
accessible sources and curated them to offer more example data sets for
use with SPATA2. If you use these data sets, please ensure that you cite
the corresponding papers correctly. When you download a
<code>SPATA2</code> object using SPATAData, the citation information is
stored in the objectâ€™s metadata, specifically in the
<code>@meta_sample</code> slot alongside all the meta-information you
find in the source data frame. You can easily obtain the citation
instructions via the <code><a href="https://rdrr.io/pkg/SPATAData/man/getCitation.html" class="external-link">getCitation()</a></code> function.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># cite Kuppe et al. 2022</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SPATAData/man/getCitation.html" class="external-link">getCitation</a></span><span class="op">(</span><span class="va">object_heart</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Citation: Kuppe, C., Ramirez Flores, R.O., Li, Z. et al. Spatial multi-omic map of human myocardial infarction. Nature 608, 766â€“777 (2022). https://doi.org/10.1038/s41586-022-05060-x</span></span>
<span><span class="co">## Source: Kuppe et al 2022</span></span></code></pre>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># cite Ravi et al. 2022</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SPATAData/man/getCitation.html" class="external-link">getCitation</a></span><span class="op">(</span><span class="va">object_gbm</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Citation: VM Ravi, P Will, J Kueckelhaus, et al. Spatially resolved multi-omics deciphers bidirectional tumor-host interdependence in glioblastoma. Cancer Cell. 2022 Jun 13;40(6):639-655.e13. doi: 10.1016/j.ccell.2022.05.009. PMID: 35700707.</span></span>
<span><span class="co">## Source: Ravi et al 2022</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="sample-meta-data">5. Sample meta data<a class="anchor" aria-label="anchor" href="#sample-meta-data"></a>
</h2>
<p>Meta data about the sample are stored in slot @<span class="citation">@meta_sample</span>. It is a list that can be extended
flexibly with <code><a href="https://rdrr.io/pkg/SPATAData/man/addSampleMetaData.html" class="external-link">addSampleMetaData()</a></code> We recommend, however, to
stick to the naming suggested by our source data.frame.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SPATAData/man/getSampleMetaData.html" class="external-link">getSampleMetaData</a></span><span class="op">(</span><span class="va">object_heart</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $sample_name</span></span>
<span><span class="co">## [1] "ACH0010"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $donor_id</span></span>
<span><span class="co">## [1] "P13"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $donor_species</span></span>
<span><span class="co">## [1] "Homo sapiens"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $histo_class</span></span>
<span><span class="co">## [1] "Cardiac Muscle"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $histo_class_sub</span></span>
<span><span class="co">##      CTRL </span></span>
<span><span class="co">## "control" </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $institution</span></span>
<span><span class="co">## [1] "University Clinic Aachen"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $lm_source</span></span>
<span><span class="co">## [1] "2024-08-24 16:13:47 CEST"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $organ</span></span>
<span><span class="co">## [1] "Heart"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $pathology</span></span>
<span><span class="co">## [1] "myocardial_infarction"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $platform</span></span>
<span><span class="co">## [1] "VisiumSmall"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $pub_citation</span></span>
<span><span class="co">## [1] "Kuppe, C., Ramirez Flores, R.O., Li, Z. et al. Spatial multi-omic map of human myocardial infarction. Nature 608, 766â€“777 (2022). https://doi.org/10.1038/s41586-022-05060-x"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $pub_doi</span></span>
<span><span class="co">## [1] "doi.org/10.1038/s41586-022-05060-x"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $pub_journal</span></span>
<span><span class="co">## [1] "Nature"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $pub_year</span></span>
<span><span class="co">## [1] 2022</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $tags</span></span>
<span><span class="co">## [1] "CTRL"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $workgroup</span></span>
<span><span class="co">## [1] "KuppeLab"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $patient_group</span></span>
<span><span class="co">## [1] "group_2"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $patient_region_id</span></span>
<span><span class="co">## [1] "GT/IZ_P13"</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="source-code-sharing">6. Source code &amp; Sharing<a class="anchor" aria-label="anchor" href="#source-code-sharing"></a>
</h2>
<p>The <code>SPATA2</code> objects have been curated manually by us
without any further processing. Data sets that derive from other
publications have been acessessed as suggested in the respective data
availability statement. <code>SPATA2</code> objects have been created in
batches as can be tracked in the script
/scripts/populate_spata2v3_objects.R in the main <a href="https://github.com/theMILOlab/SPATAData" class="external-link">repository</a> of
SPATAData. If you want to make your data set easily accessible for users
via SPATAData please contact <a href="mailto:jan.kueckelhaus@uk-erlangen.de" class="email">jan.kueckelhaus@uk-erlangen.de</a>.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Jan Kueckelhaus, Dieter-Henrik Heiland.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
